// ezdev-system/business/authorize/roleresc/Handler.js
/**
 * Created by suntf on 2017/7/24
 */
import Handler from "ezdev-pcview/handler/Handler";

export default new Handler({
    resolveTpl: function (tpl) {
        return require("./tpl/" + tpl + ".html");
    },
    initHandler:function(args){
        this.options = {
            leftId : 'leftcontent',
            rightId : 'rightcontent',
            roleListId:'roleList',
            rightTreeId:'tree_id',
            rightTreeUrl:'system/core/resource/list',
            pageId : 'pager',
            selectedRoleId : '',
            selectedRescId : '',
            lastIds : '',
            selectedRoleObj:{}
        };
    },
    initRoleList:function(){
        let args={
            contentId : "list",
            data:"search_LIKE_name=&page=1",
            form:"searchForm",
            tpl :"list",
            url :"/system/core/role/list"
        }
        let thisTemp=this;
        $('.pagination2').jqPagination({
            current_page: 1, //设置当前页 默认为1
            max_page	: 1,
            page_string : '第{current_page}页,共{max_page}页',
            paged		: function(page) {
                $("#" + this.options.leftId+' #page').val(page);

                this.searchNew(args).then(function () {
                    $("#example2 tr:first").click();
                }.bind(this));
            }.bind(this)
        });
        //this.searchRoleList();

    },
    /*新的查询方法*/
    searchNew:function(args){
        args.form = args.form || 'searchForm';
        if (args.url) {
            args.contentId = args.contentId||'list';
            let form = $(`#${args.form}`);
            if (form.length > 0) {
                args.data = form.serialize();
                args.tpl = args.tpl||'list';
                return this.render(args).then(function (data) {
                    let maxpage=parseInt(data.data.totalPages);
                    let curpage=parseInt(data.data.pageNum);
                    $(".pagination2").jqPagination("option", {
                        max_page : maxpage,
                        current_page :curpage,
                        trigger : false
                    });
                }.bind(this));
            } else {
                throw new Error("参数无效，DOM中无法找到id为[" + args.form + "]的Form！");
            }

        } else {
            throw new Error("参数无效，请传递如{url:/demo/mybatisuser/pages-(必选),type:post-(默认),tpl:list-(默认),contentId:list-(默认)}的JS对象");
        }

    },
    /*新的查询方法ed*/
    onSearchNewed:function(args){
        this.initRescTree(args);//加载资源树
        $("#example2 tr:first").click();
    },
    /**
     * 点击角色，取得授权用户，更新用户tree
     */
    initRescTree : function(args) {
        this.dialog = null;
        var tree_setting = {
            check : {
                enable : true,
                chkboxType: { "Y" : "ps", "N" : "s" }//20170925suntf
            },
            data : {
                simpleData : {
                    enable : true
                }
            },
            view :{
                fontCss: function(treeId, treeNode){
                    return (!!treeNode.highlight) ? {color:"blue", "font-weight":"bold"} : {color:"#000", "font-weight":"normal"};
                }
            }
        };
        $.ajax({
            cache : false,
            async : false,// 同步加载
            type : "POST",
            url : this.options.rightTreeUrl,
            dataType : "json",
            success : function(data) {
                $.fn.zTree.init($(`#${this.options.rightTreeId}`),tree_setting, data.data);
                var zTree = $.fn.zTree.getZTreeObj(`${this.options.rightTreeId}`);
                var root = zTree.getNodesByFilter(function (node) { return node.level == 0 }, true);
                zTree.expandNode(root, true, false, false);  //默认展开第一级
                //初始化用户树选中状态
                var element = this.options.selectedRoleObj;
                this.selRole(element);
            }.bind(this)
        });
    },
    /**
     * 点击角色，取得授权用户，更新用户tree
     */
    selRole:function(args){
        this.options.selectedRoleObj=args;
        $.fn.zTree.destroy("dataRuleTree");
        var roleid=args.objid;
        var rolename=args.objname;
        var approvalType=args.approvalType;
        //点击过后的tr变成蓝色，其他tr恢复原色
        $(`#${roleid}`).addClass("active").siblings().removeClass('active');
        this.options.selectedRoleId = roleid;
        if(!roleid){
            $("#"+this.options.rightId+" #btnSure").attr('disabled',true);
        }else{
            $("#"+this.options.rightId+" #btnSure").attr('disabled',false);
        }
        if(approvalType == "1"){
            $("#"+this.options.rightId+" #btnSure").attr('disabled',true);
            $("#"+this.options.rightId+" #btnSure").html('待审批');
        }else{
            $("#"+this.options.rightId+" #btnSure").attr('disabled',false);
            $("#"+this.options.rightId+" #btnSure").html('保存');
        }
        $("#"+this.options.rightId+" #hidRoleid").val(roleid);
        $("#"+this.options.rightId+" #spnCurRole").text(rolename);
        if(roleid){
            var url = "system/core/role/rescsbyroleid/"+roleid;
            $.ajax({
                cache : false,
                async : false,// 同步加载
                type : "GET",
                url : url,
                dataType : "json",
                success : function(data) {
                    var zTree = $.fn.zTree.getZTreeObj(this.options.rightTreeId);
                    zTree.checkAllNodes(false); //用户数选择状态初始化
                    if(data.data.length>0){
                        $.each(data.data,(key,item)=>{
                            var node = zTree.getNodeByParam("id",item.id);
                            if(node){
                                node.checked=true;
                                //zTree.updateNode(node,true);//父节点级联选中
                                zTree.updateNode(node);
                                zTree.expandNode(node.getParentNode(), true, false, false);  //展开选中节点
                            }
                        });
                    }
                    var chknodes = zTree.getCheckedNodes(true);
                    var nodes="";
                    if(chknodes.length>0){
                      $.each(chknodes,(key,item)=>{
                          if(item.type!='root'){
                              nodes+= item.id+",";
                          }
                      });
                    }
                    this.options.lastIds = nodes;
                }.bind(this)
            });
        }
    },
    /*searchRoleList:function(element){
        if(element){
            $("#" + this.options.leftId+' #page').val(1);
        }
        var data = $("#" + this.options.searchFormId).serialize();
        Utils.ajaxByPost({
            url : 'core/role/list',
            tplUrl : 'core/authorize/rolelist',
            contentId :'rolelist',
            data:data,
            calback:function(data){
                var maxpage= parseInt(data.data.totalPages);
                var curpage=parseInt(data.data.number)+ 1;
                if(maxpage==0){
                    maxpage=1;
                }
                //回写最大页数
                $(".pagination").jqPagination("option", {
                    max_page : maxpage,
                    current_page :curpage,
                    trigger : false
                });

                var main = this;
                // 左侧角色列表初始化
                $("#" + this.options.leftId + ' .authorize-role-list li').click(function() {
                    $(this).addClass('active').siblings().removeClass('active');
                    main.selRole($(this));
                });
                //初始化用户树选中状态
                var element = $("#" + this.options.leftId + ' .authorize-role-list').find('li.active');

                if(data.data.size == 0){
                    $("#"+this.options.rightId+" #btnSure").attr('disabled',true);
                }else{
                    $("#"+this.options.rightId+" #btnSure").attr('disabled',false);
                }

                main.selRole(element);
            }.bind(this)
        });
    },*/

    /*已选择*/
    selResc:function(rescid){
        this.options.selectedRescId = rescid;

    },
    bindRescs:function(args){
        var treeObj = $.fn.zTree.getZTreeObj(this.options.rightTreeId);
        var chknodes = treeObj.getCheckedNodes(true);
        var roleid= $("#"+this.options.rightId+" #hidRoleid").val();

        var nodes="";
        $.each(chknodes,(key,item)=>{
            if(item.type!='root'){
                nodes+= item.id+",";
            }
        })
        if(this.options.lastIds == nodes){
            layer.msg("绑定无变化，不可提交。");
            return;
        }

        if(roleid.length==0){
            //TODO:替换为标准提示框
            layer.msg("请选择绑定的角色！");
            return;
        }
        var param = {"roleid":roleid,"rescids":nodes};
        var url="system/core/authorize/bindrolerescs";
        $.ajax({
            cache : false,
            data:param,
            type : "POST",
            url : url,
            dataType : "json",
            success : function(data) {
                layer.msg("绑定成功！!!!");
               //this.initRescTree();绑定成功可刷新树，不刷效果一样
            }.bind(this)
        });
    },
    bindReset:function(){
        //默认选中第一个角色
        var element = this.options.selectedRoleObj;
        this.selRole(element);
    }
});
