// ezdev-system/business/datascope/Handler.js
/**
 * Created by zhaofs on 2017/5/6.
 * 演示Handler应用
 */
import Handler from "ezdev-pcview/handler/Handler";

export default new Handler({
    resolveTpl: function (tpl) {
        console.log("模板！！！！：："+tpl);
        return require("./tpl/" + tpl + ".html");
    },
    //index调用完毕
    onIndexed:function(args){

    },
    //编辑
    onEdited:function(args){
        //初始化 options
        this.options = {
            treeId:'tree_id',
            formId : 'form',
            listId : 'list',
            pageId : 'pager'
        };
        //调用生成树
        let thisObj=this;
        //setTimeout(function(){
            //this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());
            var conditionTypeBackup = $('select[name="type"]').html();
            var conditionTypeBackupTrue = $('select[name="conditionType"]').html();
            var targetBackup = $('input[name="target"]').val();

            this.targetTypeChange(conditionTypeBackup,targetBackup);

            this.valueTypeChange();

            this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());


            $('select[name="targetType"]').change(function(){

                this.targetTypeChange(conditionTypeBackup,targetBackup);
                this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());
                this.valueTypeChange();
            }.bind(this));

            $('select[name="conditionType"],select[name="type"]').change(function(){

                $('textarea[name="value"]').val('');
                this.valueTypeChange();
                this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());

            }.bind(this))
        //}.bind(this),100);
    },
    //添加
    onAdded:function(args){
        this.options = {
            treeId:'tree_id',
            formId : 'form',
            listId : 'list',
            pageId : 'pager'
        };
        //调用生成树
        //setTimeout(function(){
            //this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());
            var conditionTypeBackup = $('select[name="type"]').html();
            var conditionTypeBackupTrue = $('select[name="conditionType"]').html();
            console.log(conditionTypeBackupTrue);
            var targetBackup = $('input[name="target"]').val();

            this.targetTypeChange(conditionTypeBackup,targetBackup);

            this.valueTypeChange();

            this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());


            $('select[name="targetType"]').change(function(){

                this.targetTypeChange(conditionTypeBackup,targetBackup);
                this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());
            }.bind(this));

            $('select[name="conditionType"],select[name="type"]').change(function(){

                $('textarea[name="value"]').val('');
                this.valueTypeChange();
                this.buildEditer($('select[name="conditionType"]').val(),$('select[name="type"]').val());

            }.bind(this))
       // }.bind(this),100);

    },
    goBackB:function(args){
        $("#searchForm #zdpid").val(0);
        this.search(args);
        $("#goBackButton").hide();
    },
    //编辑时右侧树
    buildEditer: function(conditionType,valueType){
        $.fn.zTree.destroy(this.options.treeId);
        var rangeUrl;
        if(valueType == "0"){
            rangeUrl = "system/core/organization/list";

        }else if(valueType == "1"){
            rangeUrl = "system/core/user/orgusertree";

        }else{
            return;

        }

        switch (conditionType){

            case "001":case "002":case "003":case "004":case "005":case "006":case "017":
            var  tree_setting = {
                check : {
                    enable : true,
                    chkboxType: { "Y" : "", "N" : "" }
                },
                data : {
                    simpleData : {
                        enable : true
                    }
                },
                callback: {
                    onCheck: function(event, treeId, treeNode){
                        if(conditionType != "017"){
                            $('#submitForm [name="value"]').empty();
                            $('#submitForm [name="value"]').val(treeNode.id);
                        }else{

                            var treeObj = $.fn.zTree.getZTreeObj(this.options.treeId);
                            var chknodes = treeObj.getCheckedNodes(true);
                            var bindnodes="";
                            chknodes.forEach(function(item, i) {
                                bindnodes+= item.id+",";
                            });
                            $('#submitForm [name="value"]').empty();
                            $('#submitForm [name="value"]').val(bindnodes);

                        }

                    }.bind(this)
                }
            };
            if(conditionType != "017"){
                tree_setting.check.chkStyle = "radio";
                tree_setting.callback.beforeCheck =  function(treeId, treeNode){
                    if(treeNode.checked){
                        return;
                    }
                    var zTree = $.fn.zTree.getZTreeObj(this.options.treeId);
                    var nodes=zTree.getCheckedNodes(true);
                    $.each(nodes,function(i,n){
                        n.checked=false;
                    });
                    zTree.refresh();
                }.bind(this)
            }else{
                tree_setting.check.chkStyle = "checkbox";

            }

            $.ajax({
                cache : false,
                async : false,// 同步加载
                type : "POST",
                url : rangeUrl,
                dataType : "json",
                success : function(data) {
                    var nodes=data.data;
                    if(valueType == "1"){
                        nodes.forEach(function(item, i){
                            if(item.type=='user'){
                                item.iconSkin="user";

                            }else{
                                item.nocheck=true;
                            }
                        });
                    }

                    $.fn.zTree.init($('#'+this.options.treeId),tree_setting, nodes);
                    var zTree = $.fn.zTree.getZTreeObj(this.options.treeId);

                    var value = $('#submitForm [name="value"]').val();
                    if(value.indexOf(',')==-1){
                        var node = zTree.getNodeByParam("id",value);
                        if(node){
                            node.checked=true;
                            zTree.updateNode(node);
                        }
                    }else{
                        var values = value.split(',');
                        values.forEach(function(id){

                            var node = zTree.getNodeByParam("id",id);
                            if(node){
                                node.checked=true;
                                zTree.updateNode(node);
                            }
                        });
                    }
                    zTree.expandAll(true);

                }.bind(this)
            });
            break;
        }
    },
    targetTypeChange:function(conditionTypeBackup,targetBackup){

        switch ($('select[name="targetType"]').val()){
            case "0":
                $('select[name="type"]').empty().html(conditionTypeBackup);
                $('select[name="type"] option').each(function(){

                    if($(this).val() != "0" && $(this).val() != "3" && $(this).val() != "4" && $(this).val() != "5" && $(this).val() != "6"){
                        $(this).remove();
                    }
                });

                $('input[name="target"]').val('*.orgId');
                $('input[name="target"]').parents('div.form-group').hide();
                //20170915suntf【目标】非自定义时，隐藏【 自定义范围】
                $('textarea[name="value"]').parents('div.form-group').hide();
                //$('textarea[name="value"]').val('');
                $('.con-bottom-right').show();
                break;
            case "1":
                $('select[name="type"]').empty().html(conditionTypeBackup);
                $('select[name="type"] option').each(function(){
                    if($(this).val() != "1" && $(this).val() != "7"){
                        $(this).remove();
                    }
                });
                $('.con-bottom-right').show();
                $('input[name="target"]').val('*.createBy');
                $('input[name="target"]').parents('div.form-group').hide();
                //20170915suntf【目标】非自定义时，隐藏【 自定义范围】
                $('textarea[name="value"]').parents('div.form-group').hide();
                //$('textarea[name="value"]').val('');
                break;
            case "2":
                $('select[name="type"]').empty().html(conditionTypeBackup);
                $('input[name="target"]').val(targetBackup);
                $('input[name="target"]').parents('div.form-group').show();
                break;
        }
    },
    valueTypeChange: function(){

        switch ($('select[name="type"]').val()){
            case "0": case "1":
            //$('select[name="conditionType"]').removeAttr("disabled");
            $('textarea[name="value"]').parents('div.form-group').hide();
            $('.con-bottom-right').show();
            break;
            case "2":
                //$('select[name="conditionType"]').removeAttr("disabled");
                $('textarea[name="value"]').parents('div.form-group').show();
                $('.con-bottom-right').hide();
                break;
            case "3":case "7":
            $('select[name="conditionType"]').val('001')
            //$('select[name="conditionType"]').attr("disabled","disabled");
            $('textarea[name="value"]').parents('div.form-group').hide();
            $('.con-bottom-right').hide();
            break;
            case "4":case "5":case "6":
            $('select[name="conditionType"]').val('017')
            //$('select[name="conditionType"]').attr("disabled","disabled");
            $('textarea[name="value"]').parents('div.form-group').hide();
            $('.con-bottom-right').hide();
            break;
        }
    },
    checkSubmit:function (args) {
        var zTree = $.fn.zTree.getZTreeObj(this.options.treeId);
        if(zTree!=null&&zTree.getCheckedNodes().length=='0'){
            layer.msg("请选择数据范围");
        }else{
            this.submit(args);
        }
    }
});
