// ezdev-pcview/components/attach/EAttach.js
/***
 * @描述：下拉框类，继承Component
 */
import BaseComponent from "../BaseComponent";
import AjaxRequest from "../../ajax/AjaxRequest";

'use strict';
let __instance = null;
export default class EAttach extends BaseComponent {

    //覆盖BaseComponent的构造方法，修改name为e-attach
    constructor() {
        super();
        this.name = 'e-attach';
        this.mimeTbl = {
            "0123":"application/vndlotus-1-2-3",
            "3dml":"text/vndin3d3dml",
            "3g2":"video/3gpp2",
            "3gp":"video/3gpp",
            "7z":"application/x-7z-compressed",
            "aab":"application/x-authorware-bin",
            "aac":"audio/x-aac",
            "aam":"application/x-authorware-map",
            "aas":"application/x-authorware-seg",
            "abw":"application/x-abiword",
            "ac":"application/pkix-attr-cert",
            "acc":"application/vndamericandynamicsacc",
            "ace":"application/x-ace-compressed",
            "acu":"application/vndacucobol",
            "adp":"audio/adpcm",
            "aep":"application/vndaudiograph",
            "afp":"application/vndibmmodcap",
            "ahead":"application/vndaheadspace",
            "ai":"application/postscript",
            "aif":"audio/x-aiff",
            "air":"application/vndadobeair-application-installer-package+zip",
            "ait":"application/vnddvbait",
            "ami":"application/vndamigaami",
            "apk":"application/vndandroidpackage-archive",
            "application":"application/x-ms-application",
            "apr":"application/vndlotus-approach",
            "asf":"video/x-ms-asf",
            "aso":"application/vndaccpacsimplyaso",
            "atc":"application/vndacucorp",
            "atom, xml":"application/atom+xml",
            "atomcat":"application/atomcat+xml",
            "atomsvc":"application/atomsvc+xml",
            "atx":"application/vndantixgame-component",
            "au":"audio/basic",
            "avi":"video/x-msvideo",
            "aw":"application/applixware",
            "azf":"application/vndairzipfilesecureazf",
            "azs":"application/vndairzipfilesecureazs",
            "azw":"application/vndamazonebook",
            "bcpio":"application/x-bcpio",
            "bdf":"application/x-font-bdf",
            "bdm":"application/vndsyncmldm+wbxml",
            "bed":"application/vndrealvncbed",
            "bh2":"application/vndfujitsuoasysprs",
            "bin":"application/octet-stream",
            "bmi":"application/vndbmi",
            "bmp":"image/bmp",
            "box":"application/vndpreviewsystemsbox",
            "btif":"image/prsbtif",
            "bz":"application/x-bzip",
            "bz2":"application/x-bzip2",
            "c":"text/x-c",
            "c11amc":"application/vndcluetrustcartomobile-config",
            "c11amz":"application/vndcluetrustcartomobile-config-pkg",
            "c4g":"application/vndclonkc4group",
            "cab":"application/vndms-cab-compressed",
            "car":"application/vndcurlcar",
            "cat":"application/vndms-pkiseccat",
            "ccxml":"application/ccxml+xml,",
            "cdbcmsg":"application/vndcontactcmsg",
            "cdkey":"application/vndmediastationcdkey",
            "cdmia":"application/cdmi-capability",
            "cdmic":"application/cdmi-container",
            "cdmid":"application/cdmi-domain",
            "cdmio":"application/cdmi-object",
            "cdmiq":"application/cdmi-queue",
            "cdx":"chemical/x-cdx",
            "cdxml":"application/vndchemdraw+xml",
            "cdy":"application/vndcinderella",
            "cer":"application/pkix-cert",
            "cgm":"image/cgm",
            "chat":"application/x-chat",
            "chm":"application/vndms-htmlhelp",
            "chrt":"application/vndkdekchart",
            "cif":"chemical/x-cif",
            "cii":"application/vndanser-web-certificate-issue-initiation",
            "cil":"application/vndms-artgalry",
            "cla":"application/vndclaymore",
            "class":"application/java-vm",
            "clkk":"application/vndcrickclickerkeyboard",
            "clkp":"application/vndcrickclickerpalette",
            "clkt":"application/vndcrickclickertemplate",
            "clkw":"application/vndcrickclickerwordbank",
            "clkx":"application/vndcrickclicker",
            "clp":"application/x-msclip",
            "cmc":"application/vndcosmocaller",
            "cmdf":"chemical/x-cmdf",
            "cml":"chemical/x-cml",
            "cmp":"application/vndyellowriver-custom-menu",
            "cmx":"image/x-cmx",
            "cod":"application/vndrimcod",
            "cpio":"application/x-cpio",
            "cpt":"application/mac-compactpro",
            "crd":"application/x-mscardfile",
            "crl":"application/pkix-crl",
            "cryptonote":"application/vndrigcryptonote",
            "csh":"application/x-csh",
            "csml":"chemical/x-csml",
            "csp":"application/vndcommonspace",
            "css":"text/css",
            "csv":"text/csv",
            "cu":"application/cu-seeme",
            "curl":"text/vndcurl",
            "cww":"application/prscww",
            "dae":"model/vndcollada+xml",
            "daf":"application/vndmobiusdaf",
            "davmount":"application/davmount+xml",
            "dcurl":"text/vndcurldcurl",
            "dd2":"application/vndomadd2+xml",
            "ddd":"application/vndfujixeroxddd",
            "deb":"application/x-debian-package",
            "der":"application/x-x509-ca-cert",
            "dfac":"application/vnddreamfactory",
            "dir":"application/x-director",
            "dis":"application/vndmobiusdis",
            "djvu":"image/vnddjvu",
            "dna":"application/vnddna",
            "doc":"application/msword",
            "docm":"application/vndms-worddocumentmacroenabled12",
            "docx":"application/vndopenxmlformats-officedocumentwordprocessingmldocument",
            "dotm":"application/vndms-wordtemplatemacroenabled12",
            "dotx":"application/vndopenxmlformats-officedocumentwordprocessingmltemplate",
            "dp":"application/vndosgidp",
            "dpg":"application/vnddpgraph",
            "dra":"audio/vnddra",
            "dsc":"text/prslinestag",
            "dssc":"application/dssc+der",
            "dtb":"application/x-dtbook+xml",
            "dtd":"application/xml-dtd",
            "dts":"audio/vnddts",
            "dtshd":"audio/vnddtshd",
            "dvi":"application/x-dvi",
            "dwf":"model/vnddwf",
            "dwg":"image/vnddwg",
            "dxf":"image/vnddxf",
            "dxp":"application/vndspotfiredxp",
            "ecelp4800":"audio/vndnueraecelp4800",
            "ecelp7470":"audio/vndnueraecelp7470",
            "ecelp9600":"audio/vndnueraecelp9600",
            "edm":"application/vndnovadigmedm",
            "edx":"application/vndnovadigmedx",
            "efif":"application/vndpicsel",
            "ei6":"application/vndpgosasli",
            "eml":"message/rfc822",
            "emma":"application/emma+xml",
            "eol":"audio/vnddigital-winds",
            "eot":"application/vndms-fontobject",
            "epub":"application/epub+zip",
            "es":"application/ecmascript",
            "es3":"application/vndeszigno3+xml",
            "esf":"application/vndepsonesf",
            "etx":"text/x-setext",
            "exe":"application/x-msdownload",
            "exi":"application/exi",
            "ext":"application/vndnovadigmext",
            "ez2":"application/vndezpix-album",
            "ez3":"application/vndezpix-package",
            "f":"text/x-fortran",
            "f4v":"video/x-f4v",
            "fbs":"image/vndfastbidsheet",
            "fcs":"application/vndisacfcs",
            "fdf":"application/vndfdf",
            "fe_launch":"application/vnddenovofcselayout-link",
            "fg5":"application/vndfujitsuoasysgp",
            "fh":"image/x-freehand",
            "fig":"application/x-xfig",
            "fli":"video/x-fli",
            "flo":"application/vndmicrografxflo",
            "flv":"video/x-flv",
            "flw":"application/vndkdekivio",
            "flx":"text/vndfmiflexstor",
            "fly":"text/vndfly",
            "fm":"application/vndframemaker",
            "fnc":"application/vndfrogansfnc",
            "fpx":"image/vndfpx",
            "fsc":"application/vndfscweblaunch",
            "fst":"image/vndfst",
            "ftc":"application/vndfluxtimeclip",
            "fti":"application/vndanser-web-funds-transfer-initiation",
            "fvt":"video/vndfvt",
            "fxp":"application/vndadobefxp",
            "fzs":"application/vndfuzzysheet",
            "g2w":"application/vndgeoplan",
            "g3":"image/g3fax",
            "g3w":"application/vndgeospace",
            "gac":"application/vndgroove-account",
            "gdl":"model/vndgdl",
            "geo":"application/vnddynageo",
            "gex":"application/vndgeometry-explorer",
            "ggb":"application/vndgeogebrafile",
            "ggt":"application/vndgeogebratool",
            "ghf":"application/vndgroove-help",
            "gif":"image/gif",
            "gim":"application/vndgroove-identity-message",
            "gmx":"application/vndgmx",
            "gnumeric":"application/x-gnumeric",
            "gph":"application/vndflographit",
            "gqf":"application/vndgrafeq",
            "gram":"application/srgs",
            "grv":"application/vndgroove-injector",
            "grxml":"application/srgs+xml",
            "gsf":"application/x-font-ghostscript",
            "gtar":"application/x-gtar",
            "gtm":"application/vndgroove-tool-message",
            "gtw":"model/vndgtw",
            "gv":"text/vndgraphviz",
            "gxt":"application/vndgeonext",
            "h261":"video/h261",
            "h263":"video/h263",
            "h264":"video/h264",
            "hal":"application/vndhal+xml",
            "hbci":"application/vndhbci",
            "hdf":"application/x-hdf",
            "hlp":"application/winhlp",
            "hpgl":"application/vndhp-hpgl",
            "hpid":"application/vndhp-hpid",
            "hps":"application/vndhp-hps",
            "hqx":"application/mac-binhex40",
            "htke":"application/vndkenameaapp",
            "html":"text/html",
            "hvd":"application/vndyamahahv-dic",
            "hvp":"application/vndyamahahv-voice",
            "hvs":"application/vndyamahahv-script",
            "i2g":"application/vndintergeo",
            "icc":"application/vndiccprofile",
            "ice":"x-conference/x-cooltalk",
            "ico":"image/x-icon",
            "ics":"text/calendar",
            "ief":"image/ief",
            "ifm":"application/vndshanainformedformdata",
            "igl":"application/vndigloader",
            "igm":"application/vndinsorsigm",
            "igs":"model/iges",
            "igx":"application/vndmicrografxigx",
            "iif":"application/vndshanainformedinterchange",
            "imp":"application/vndaccpacsimplyimp",
            "ims":"application/vndms-ims",
            "ipfix":"application/ipfix",
            "ipk":"application/vndshanainformedpackage",
            "irm":"application/vndibmrights-management",
            "irp":"application/vndirepositorypackage+xml",
            "itp":"application/vndshanainformedformtemplate",
            "ivp":"application/vndimmervision-ivp",
            "ivu":"application/vndimmervision-ivu",
            "jad":"text/vndsunj2meapp-descriptor",
            "jam":"application/vndjam",
            "jar":"application/java-archive",
            "java":"text/x-java-source,java",
            "jisp":"application/vndjisp",
            "jlt":"application/vndhp-jlyt",
            "jnlp":"application/x-java-jnlp-file",
            "joda":"application/vndjoostjoda-archive",
            "jpeg":"image/jpeg",
            "jpg":"image/jpeg",
            "jpgv":"video/jpeg",
            "jpm":"video/jpm",
            "js":"application/javascript",
            "json":"application/json",
            "karbon":"application/vndkdekarbon",
            "kfo":"application/vndkdekformula",
            "kia":"application/vndkidspiration",
            "kml":"application/vndgoogle-earthkml+xml",
            "kmz":"application/vndgoogle-earthkmz",
            "kne":"application/vndkinar",
            "kon":"application/vndkdekontour",
            "kpr":"application/vndkdekpresenter",
            "ksp":"application/vndkdekspread",
            "ktx":"image/ktx",
            "ktz":"application/vndkahootz",
            "kwd":"application/vndkdekword",
            "lasxml":"application/vndlaslas+xml",
            "latex":"application/x-latex",
            "lbd":"application/vndllamagraphicslife-balancedesktop",
            "lbe":"application/vndllamagraphicslife-balanceexchange+xml",
            "les":"application/vndhhelesson-player",
            "link66":"application/vndroute66link66+xml",
            "lrm":"application/vndms-lrm",
            "ltf":"application/vndfrogansltf",
            "lvp":"audio/vndlucentvoice",
            "lwp":"application/vndlotus-wordpro",
            "m21":"application/mp21",
            "m3u":"audio/x-mpegurl",
            "m3u8":"application/vndapplempegurl",
            "m4v":"video/x-m4v",
            "ma":"application/mathematica",
            "mads":"application/mads+xml",
            "mag":"application/vndecowinchart",
            "mathml":"application/mathml+xml",
            "mbk":"application/vndmobiusmbk",
            "mbox":"application/mbox",
            "mc1":"application/vndmedcalcdata",
            "mcd":"application/vndmcd",
            "mcurl":"text/vndcurlmcurl",
            "mdb":"application/x-msaccess",
            "mdi":"image/vndms-modi",
            "meta4":"application/metalink4+xml",
            "mets":"application/mets+xml",
            "mfm":"application/vndmfmp",
            "mgp":"application/vndosgeomapguidepackage",
            "mgz":"application/vndproteusmagazine",
            "mid":"audio/midi",
            "mif":"application/vndmif",
            "mj2":"video/mj2",
            "mlp":"application/vnddolbymlp",
            "mmd":"application/vndchipnutskaraoke-mmd",
            "mmf":"application/vndsmaf",
            "mmr":"image/vndfujixeroxedmics-mmr",
            "mny":"application/x-msmoney",
            "mods":"application/mods+xml",
            "movie":"video/x-sgi-movie",
            "mp2":"audio/mpeg",
            "mp3":"audio/mpeg",
            "mp4":"video/mp4",
            "mp4a":"audio/mp4",
            "mpc":"application/vndmophuncertificate",
            "mpeg":"video/mpeg",
            "mpga":"audio/mpeg",
            "mpkg":"application/vndappleinstaller+xml",
            "mpm":"application/vndblueicemultipass",
            "mpn":"application/vndmophunapplication",
            "mpp":"application/vndms-project",
            "mpy":"application/vndibmminipay",
            "mqy":"application/vndmobiusmqy",
            "mrc":"application/marc",
            "mrcx":"application/marcxml+xml",
            "mscml":"application/mediaservercontrol+xml",
            "mseq":"application/vndmseq",
            "msf":"application/vndepsonmsf",
            "msh":"model/mesh",
            "msl":"application/vndmobiusmsl",
            "msty":"application/vndmuveestyle",
            "mts":"model/vndmts",
            "mus":"application/vndmusician",
            "musicxml":"application/vndrecordaremusicxml+xml",
            "mvb":"application/x-msmediaview",
            "mwf":"application/vndmfer",
            "mxf":"application/mxf",
            "mxl":"application/vndrecordaremusicxml",
            "mxml":"application/xv+xml",
            "mxs":"application/vndtriscapemxs",
            "mxu":"video/vndmpegurl",
            "n3":"text/n3",
            "nbp":"application/vndwolframplayer",
            "nc":"application/x-netcdf",
            "ncx":"application/x-dtbncx+xml",
            "n-gage":"application/vndnokian-gagesymbianinstall",
            "ngdat":"application/vndnokian-gagedata",
            "nlu":"application/vndneurolanguagenlu",
            "nml":"application/vndenliven",
            "nnd":"application/vndnoblenet-directory",
            "nns":"application/vndnoblenet-sealer",
            "nnw":"application/vndnoblenet-web",
            "npx":"image/vndnet-fpx",
            "nsf":"application/vndlotus-notes",
            "oa2":"application/vndfujitsuoasys2",
            "oa3":"application/vndfujitsuoasys3",
            "oas":"application/vndfujitsuoasys",
            "obd":"application/x-msbinder",
            "oda":"application/oda",
            "odb":"application/vndoasisopendocumentdatabase",
            "odc":"application/vndoasisopendocumentchart",
            "odf":"application/vndoasisopendocumentformula",
            "odft":"application/vndoasisopendocumentformula-template",
            "odg":"application/vndoasisopendocumentgraphics",
            "odi":"application/vndoasisopendocumentimage",
            "odm":"application/vndoasisopendocumenttext-master",
            "odp":"application/vndoasisopendocumentpresentation",
            "ods":"application/vndoasisopendocumentspreadsheet",
            "odt":"application/vndoasisopendocumenttext",
            "oga":"audio/ogg",
            "ogv":"video/ogg",
            "ogx":"application/ogg",
            "onetoc":"application/onenote",
            "opf":"application/oebps-package+xml",
            "org":"application/vndlotus-organizer",
            "osf":"application/vndyamahaopenscoreformat",
            "osfpvg":"application/vndyamahaopenscoreformatosfpvg+xml",
            "otc":"application/vndoasisopendocumentchart-template",
            "otf":"application/x-font-otf",
            "otg":"application/vndoasisopendocumentgraphics-template",
            "oth":"application/vndoasisopendocumenttext-web",
            "oti":"application/vndoasisopendocumentimage-template",
            "otp":"application/vndoasisopendocumentpresentation-template",
            "ots":"application/vndoasisopendocumentspreadsheet-template",
            "ott":"application/vndoasisopendocumenttext-template",
            "oxt":"application/vndopenofficeorgextension",
            "p":"text/x-pascal",
            "p10":"application/pkcs10",
            "p12":"application/x-pkcs12",
            "p7b":"application/x-pkcs7-certificates",
            "p7m":"application/pkcs7-mime",
            "p7r":"application/x-pkcs7-certreqresp",
            "p7s":"application/pkcs7-signature",
            "p8":"application/pkcs8",
            "par":"text/plain-bas",
            "paw":"application/vndpawaafile",
            "pbd":"application/vndpowerbuilder6",
            "pbm":"image/x-portable-bitmap",
            "pcf":"application/x-font-pcf",
            "pcl":"application/vndhp-pcl",
            "pclxl":"application/vndhp-pclxl",
            "pcurl":"application/vndcurlpcurl",
            "pcx":"image/x-pcx",
            "pdb":"application/vndpalm",
            "pdf":"application/pdf",
            "pfa":"application/x-font-type1",
            "pfr":"application/font-tdpfr",
            "pgm":"image/x-portable-graymap",
            "pgn":"application/x-chess-pgn",
            "pgp":"application/pgp-encrypted",
            "pic":"image/x-pict",
            "pjpeg":"image/pjpeg",
            "pki":"application/pkixcmp",
            "pkipath":"application/pkix-pkipath",
            "plb":"application/vnd3gpppic-bw-large",
            "plc":"application/vndmobiusplc",
            "plf":"application/vndpocketlearn",
            "pls":"application/pls+xml",
            "pml":"application/vndctc-posml",
            "png":"image/png",
            "pnm":"image/x-portable-anymap",
            "portpkg":"application/vndmacportsportpkg",
            "potm":"application/vndms-powerpointtemplatemacroenabled12",
            "potx":"application/vndopenxmlformats-officedocumentpresentationmltemplate",
            "ppam":"application/vndms-powerpointaddinmacroenabled12",
            "ppd":"application/vndcups-ppd",
            "ppm":"image/x-portable-pixmap",
            "ppsm":"application/vndms-powerpointslideshowmacroenabled12",
            "ppsx":"application/vndopenxmlformats-officedocumentpresentationmlslideshow",
            "ppt":"application/vndms-powerpoint",
            "pptm":"application/vndms-powerpointpresentationmacroenabled12",
            "pptx":"application/vndopenxmlformats-officedocumentpresentationmlpresentation",
            "prc":"application/x-mobipocket-ebook",
            "pre":"application/vndlotus-freelance",
            "prf":"application/pics-rules",
            "psd":"image/vndadobephotoshop",
            "psf":"application/x-font-linux-psf",
            "pskcxml":"application/pskc+xml",
            "ptid":"application/vndpviptid1",
            "pub":"application/x-mspublisher",
            "pvb":"application/vnd3gpppic-bw-var",
            "pwn":"application/vnd3mpost-it-notes",
            "pya":"audio/vndms-playreadymediapya",
            "pyv":"video/vndms-playreadymediapyv",
            "qam":"application/vndepsonquickanime",
            "qbo":"application/vndintuqbo",
            "qfx":"application/vndintuqfx",
            "qps":"application/vndpublishare-delta-tree",
            "qt":"video/quicktime",
            "qxd":"application/vndquarkquarkxpress",
            "ram":"audio/x-pn-realaudio",
            "rar":"application/x-rar-compressed",
            "ras":"image/x-cmu-raster",
            "rcprofile":"application/vndipunpluggedrcprofile",
            "rdf":"application/rdf+xml",
            "rdz":"application/vnddata-visionrdz",
            "rep":"application/vndbusinessobjects",
            "res":"application/x-dtbresource+xml",
            "rgb":"image/x-rgb",
            "rif":"application/reginfo+xml",
            "rip":"audio/vndrip",
            "rl":"application/resource-lists+xml",
            "rlc":"image/vndfujixeroxedmics-rlc",
            "rld":"application/resource-lists-diff+xml",
            "rm":"application/vndrn-realmedia",
            "rmp":"audio/x-pn-realaudio-plugin",
            "rms":"application/vndjcpjavamemidlet-rms",
            "rnc":"application/relax-ng-compact-syntax",
            "rp9":"application/vndcloantorp9",
            "rpss":"application/vndnokiaradio-presets",
            "rpst":"application/vndnokiaradio-preset",
            "rq":"application/sparql-query",
            "rs":"application/rls-services+xml",
            "rsd":"application/rsd+xml",
            "rss, xml":"application/rss+xml",
            "rtf":"application/rtf",
            "rtx":"text/richtext",
            "s":"text/x-asm",
            "saf":"application/vndyamahasmaf-audio",
            "sbml":"application/sbml+xml",
            "sc":"application/vndibmsecure-container",
            "scd":"application/x-msschedule",
            "scm":"application/vndlotus-screencam",
            "scq":"application/scvp-cv-request",
            "scs":"application/scvp-cv-response",
            "scurl":"text/vndcurlscurl",
            "sda":"application/vndstardivisiondraw",
            "sdc":"application/vndstardivisioncalc",
            "sdd":"application/vndstardivisionimpress",
            "sdkm":"application/vndsolentsdkm+xml",
            "sdp":"application/sdp",
            "sdw":"application/vndstardivisionwriter",
            "see":"application/vndseemail",
            "seed":"application/vndfdsnseed",
            "sema":"application/vndsema",
            "semd":"application/vndsemd",
            "semf":"application/vndsemf",
            "ser":"application/java-serialized-object",
            "setpay":"application/set-payment-initiation",
            "setreg":"application/set-registration-initiation",
            "sfd-hdstx":"application/vndhydrostatixsof-data",
            "sfs":"application/vndspotfiresfs",
            "sgl":"application/vndstardivisionwriter-global",
            "sgml":"text/sgml",
            "sh":"application/x-sh",
            "shar":"application/x-shar",
            "shf":"application/shf+xml",
            "sis":"application/vndsymbianinstall",
            "sit":"application/x-stuffit",
            "sitx":"application/x-stuffitx",
            "skp":"application/vndkoan",
            "sldm":"application/vndms-powerpointslidemacroenabled12",
            "sldx":"application/vndopenxmlformats-officedocumentpresentationmlslide",
            "slt":"application/vndepsonsalt",
            "sm":"application/vndstepmaniastepchart",
            "smf":"application/vndstardivisionmath",
            "smi":"application/smil+xml",
            "snf":"application/x-font-snf",
            "spf":"application/vndyamahasmaf-phrase",
            "spl":"application/x-futuresplash",
            "spot":"text/vndin3dspot",
            "spp":"application/scvp-vp-response",
            "spq":"application/scvp-vp-request",
            "src":"application/x-wais-source",
            "sru":"application/sru+xml",
            "srx":"application/sparql-results+xml",
            "sse":"application/vndkodak-descriptor",
            "ssf":"application/vndepsonssf",
            "ssml":"application/ssml+xml",
            "st":"application/vndsailingtrackertrack",
            "stc":"application/vndsunxmlcalctemplate",
            "std":"application/vndsunxmldrawtemplate",
            "stf":"application/vndwtstf",
            "sti":"application/vndsunxmlimpresstemplate",
            "stk":"application/hyperstudio",
            "stl":"application/vndms-pkistl",
            "str":"application/vndpgformat",
            "stw":"application/vndsunxmlwritertemplate",
            "sub":"image/vnddvbsubtitle",
            "sus":"application/vndsus-calendar",
            "sv4cpio":"application/x-sv4cpio",
            "sv4crc":"application/x-sv4crc",
            "svc":"application/vnddvbservice",
            "svd":"application/vndsvd",
            "svg":"image/svg+xml",
            "swf":"application/x-shockwave-flash",
            "swi":"application/vndaristanetworksswi",
            "sxc":"application/vndsunxmlcalc",
            "sxd":"application/vndsunxmldraw",
            "sxg":"application/vndsunxmlwriterglobal",
            "sxi":"application/vndsunxmlimpress",
            "sxm":"application/vndsunxmlmath",
            "sxw":"application/vndsunxmlwriter",
            "t":"text/troff",
            "tao":"application/vndtaointent-module-archive",
            "tar":"application/x-tar",
            "tcap":"application/vnd3gpp2tcap",
            "tcl":"application/x-tcl",
            "teacher":"application/vndsmartteacher",
            "tei":"application/tei+xml",
            "tex":"application/x-tex",
            "texinfo":"application/x-texinfo",
            "tfi":"application/thraud+xml",
            "tfm":"application/x-tex-tfm",
            "thmx":"application/vndms-officetheme",
            "tiff":"image/tiff",
            "tmo":"application/vndtmobile-livetv",
            "torrent":"application/x-bittorrent",
            "tpl":"application/vndgroove-tool-template",
            "tpt":"application/vndtridtpt",
            "tra":"application/vndtrueapp",
            "trm":"application/x-msterminal",
            "tsd":"application/timestamped-data",
            "tsv":"text/tab-separated-values",
            "ttf":"application/x-font-ttf",
            "ttl":"text/turtle",
            "twd":"application/vndsimtech-mindmapper",
            "txd":"application/vndgenomatixtuxedo",
            "txf":"application/vndmobiustxf",
            "txt":"text/plain",
            "ufd":"application/vndufdl",
            "umj":"application/vndumajin",
            "unityweb":"application/vndunity",
            "uoml":"application/vnduoml+xml",
            "uri":"text/uri-list",
            "ustar":"application/x-ustar",
            "utz":"application/vnduiqtheme",
            "uu":"text/x-uuencode",
            "uva":"audio/vnddeceaudio",
            "uvh":"video/vnddecehd",
            "uvi":"image/vnddecegraphic",
            "uvm":"video/vnddecemobile",
            "uvp":"video/vnddecepd",
            "uvs":"video/vnddecesd",
            "uvu":"video/vnduvvump4",
            "uvv":"video/vnddecevideo",
            "vcd":"application/x-cdlink",
            "vcf":"text/x-vcard",
            "vcg":"application/vndgroove-vcard",
            "vcs":"text/x-vcalendar",
            "vcx":"application/vndvcx",
            "vis":"application/vndvisionary",
            "viv":"video/vndvivo",
            "vsd":"application/vndvisio",
            "vsdx":"application/vndvisio2013",
            "vsf":"application/vndvsf",
            "vtu":"model/vndvtu",
            "vxml":"application/voicexml+xml",
            "wad":"application/x-doom",
            "wav":"audio/x-wav",
            "wax":"audio/x-ms-wax",
            "wbmp":"image/vndwapwbmp",
            "wbs":"application/vndcriticaltoolswbs+xml",
            "wbxml":"application/vndwapwbxml",
            "weba":"audio/webm",
            "webm":"video/webm",
            "webp":"image/webp",
            "wg":"application/vndpmiwidget",
            "wgt":"application/widget",
            "wm":"video/x-ms-wm",
            "wma":"audio/x-ms-wma",
            "wmd":"application/x-ms-wmd",
            "wmf":"application/x-msmetafile",
            "wml":"text/vndwapwml",
            "wmlc":"application/vndwapwmlc",
            "wmls":"text/vndwapwmlscript",
            "wmlsc":"application/vndwapwmlscriptc",
            "wmv":"video/x-ms-wmv",
            "wmx":"video/x-ms-wmx",
            "wmz":"application/x-ms-wmz",
            "woff":"application/x-font-woff",
            "wpd":"application/vndwordperfect",
            "wpl":"application/vndms-wpl",
            "wps":"application/vndms-works",
            "wqd":"application/vndwqd",
            "wri":"application/x-mswrite",
            "wrl":"model/vrml",
            "wsdl":"application/wsdl+xml",
            "wspolicy":"application/wspolicy+xml",
            "wtb":"application/vndwebturbo",
            "wvx":"video/x-ms-wvx",
            "x3d":"application/vndhzn-3d-crossword",
            "xap":"application/x-silverlight-app",
            "xar":"application/vndxara",
            "xbap":"application/x-ms-xbap",
            "xbd":"application/vndfujixeroxdocuworksbinder",
            "xbm":"image/x-xbitmap",
            "xdf":"application/xcap-diff+xml",
            "xdm":"application/vndsyncmldm+xml",
            "xdp":"application/vndadobexdp+xml",
            "xdssc":"application/dssc+xml",
            "xdw":"application/vndfujixeroxdocuworks",
            "xenc":"application/xenc+xml",
            "xer":"application/patch-ops-error+xml",
            "xfdf":"application/vndadobexfdf",
            "xfdl":"application/vndxfdl",
            "xhtml":"application/xhtml+xml",
            "xif":"image/vndxiff",
            "xlam":"application/vndms-exceladdinmacroenabled12",
            "xls":"application/vndms-excel",
            "xlsb":"application/vndms-excelsheetbinarymacroenabled12",
            "xlsm":"application/vndms-excelsheetmacroenabled12",
            "xlsx":"application/vndopenxmlformats-officedocumentspreadsheetmlsheet",
            "xltm":"application/vndms-exceltemplatemacroenabled12",
            "xltx":"application/vndopenxmlformats-officedocumentspreadsheetmltemplate",
            "xml":"application/xml",
            "xo":"application/vndolpc-sugar",
            "xop":"application/xop+xml",
            "xpi":"application/x-xpinstall",
            "xpm":"image/x-xpixmap",
            "xpr":"application/vndis-xpr",
            "xps":"application/vndms-xpsdocument",
            "xpw":"application/vndinterconformnet",
            "xslt":"application/xslt+xml",
            "xsm":"application/vndsyncml+xml",
            "xspf":"application/xspf+xml",
            "xul":"application/vndmozillaxul+xml",
            "xwd":"image/x-xwindowdump",
            "xyz":"chemical/x-xyz",
            "yaml":"text/yaml",
            "yang":"application/yang",
            "yin":"application/yin+xml",
            "zaz":"application/vndzzazzdeck+xml",
            "zip":"application/zip",
            "zir":"application/vndzul",
            "zmm":"application/vndhandheld-entertainment+xml"
        }
    }

    //覆盖BaseComponent的instance方法，提供EAttach实例
    static instance() {
        if (__instance == null) {
            __instance = new EAttach();
        }
        return __instance;
    }

    /**
     * 根据<e-attach>元素的属性，生成必要的config内容，设置到参数内。
     */
    setArgsBeforeRw(args, options) {

        //把属性放入args里，一般控件的基础动作。
        args = super.setArgsBeforeRw(args, options);

        //如果使用者自定义了attachmentId，把它作为附件的唯一标识；如果未定义，生成。
        args.attachmentId = args.attachmentId ? args.attachmentId : this.guid();
        //如果没有设定type，默认为default
        args.atrType = (!args.atrType)?"default":args.atrType;

        //如果没有设定urlBase，默认为/attach
        args.attachUrlBase = args.attachUrlBase ? args.attachUrlBase : "/attach";
        //如果没有设定url，默认为/attach/list/id
        args.url = args.url ? args.url : args.attachUrlBase + "/list/" + args.attachmentId;


        //回显还是初显
        if(args.atrType.indexOf("showback-") == 0){
            args.atrType = args.atrType.substr(9);
            args.isShowBack = true;
        }

        //清除原页面，准备重新生成。
        if($('#' + args.attachmentId)[0]){
            $('#' + args.attachmentId).remove();
        }

        return args;
    }

    /**
     * 生成e-attach的html
     */
    generateHtml(args) {

        var attachmentId = args.attachmentId;

        //获取后台数据
        this.refreshDatas(args);

        //生成attacher的主div
        var attacher = $('<div id="atr-main-'+attachmentId+'" class="wu-example" ></div>');

        //生成存放attachmentId的元素
        var idSaver = $('<input type="hidden" id="'+args.atrIdSaver+'" value="'+attachmentId+'" />');
        var nameSaver = $('<input type="hidden" name="'+args.atrNameSaver+'" value="'+attachmentId+'" />');
        attacher.append(idSaver);
        attacher.append(nameSaver);

        //初始化显示
        if(args.atrType=="default"){
            if(!args.isShowBack)
                attacher.append(this.getInputRow(args, "default"));
            attacher.append(this.getReadonlyInputRow(args));
        }else if(args.atrType=="aLink"){
            if(!args.isShowBack)
                attacher.append(this.getInputRow(args, "aLink"));
            attacher.append(this.getALinkRow(args));
        }else if(args.atrType=="img42"){
            if(!args.isShowBack)
                attacher.append(this.getInputRow(args, "img42"));
            attacher.append(this.getImgRow(args));
        }else{
            attacher.append($('<input value="请检查e-attach控件的atrType属性值。" style="width:300px" readonly="readonly">'));
        }

        //整个节点的html
        return attacher.prop("outerHTML");
    }

    /**
     * e-attach控件生成后的WebUploader注册操作
     */
    executeAferRw(args) {

        var attachmentId = args.attachmentId;
        var $ = jQuery;

        //为刚生成的控件注册事件
        this.registEvents(args);

        //为图片类型的回显，注册图片切换器
        var swiper = new Swiper('.swiper-container', {//图片切换器
            zoom: true,
            pagination: '.swiper-pagination',
            nextButton: '.swiper-button-next',
            prevButton: '.swiper-button-prev',
            slidesPerView: 5,
            spaceBetween: 10,
            preventClicks : false
        });

        if(args.isShowBack){
            return;
        }

        //后缀限制
        var acceptObj = null;
        if(args.suffix){
            var mimeTypes = this.getMime(args.suffix);
            if(mimeTypes){
                acceptObj = {
                    title:"自定义后缀限制", extensions:args.suffix, mimeTypes:mimeTypes
                }
            }
        }

        //自定义上传url
        var uploadUrl = '/attach/upload/'+attachmentId;
        if(args.uploadUrl){
            uploadUrl = args.uploadUrl;
        }

        //自定义分片
        var chunked = args.chunked ? args.chunked : true;

        //自定义（像素格式、单/多附件、带全文检索功能）参数
        var formData = {};
        var pixelFlg = args.pixels?args.pixels:"original";
        formData.pixels=pixelFlg;
        if(args.single && args.single=="true"){
            formData.single="true";
        }
        if(args.searchDoc && args.searchDoc==true){
            formData.searchDoc="true";
        }

        var uploader;//上传器
        // WebUploader的全局hooks
        WebUploader.Uploader.register({
            "before-send-file": "beforeSendFile",  // 整个文件上传前
            "before-send": "beforeSend",           // 每个分片上传前
            "after-send-file": "afterSendFile"     // 分片上传完毕
        }, {
            beforeSendFile: function (file) {
                var owner = this.owner;
                var deferred = WebUploader.Base.Deferred();
                formData.fileMd5 = undefined;//重置md5
                console.log("generating md5");
                formData.md5LoadNoNeedShow = false;
                layer.msg('正在计算已上传进度，请稍候。', {
                    icon: 6,
                    time: 3000
                });
                if(!formData.md5LoadNoNeedShow) {
                    formData.md5LoadIndex = layer.load(2);
                }

                // //获得标准HTML5的文件
                // var blob = uploader.getFile(file).source;
                // var blobMd5 = this.getMd5(blob.source);
                // console.log("[独立计算MD5]："+blobMd5);

                uploader.md5File( file )
                    .progress(function(percentage) {
                        console.log('Percentage:', percentage);
                    })
                    .fail(function() {
                        layer.close(formData.md5LoadIndex);
                        formData.md5LoadNoNeedShow = true;
                        alert("计算上传进度(本地文件md5值)失败。");
                        deferred.reject();
                    })
                    .then(function(val) {
                        console.log('generated md5 result:', val);
                        formData.fileMd5 = val;
                        uploader.option('formData', formData);
                        //获取已上传的分片的MD5
                        var chunkInfoData = {};
                        chunkInfoData.attachmentId = attachmentId;
                        chunkInfoData.fileName = file.name;
                        chunkInfoData.fileMd5 = val;
                        AjaxRequest.ajax(
                            {
                                url:args.attachUrlBase+"/chunkInfo",
                                data:chunkInfoData
                            })
                            .then(function(successData){
                                    formData.uploadedChunksMd5 = successData.data;
                                    uploader.option( 'formData', formData);
                                    deferred.resolve();
                                }.bind(this),
                                function(failData){
                                    if(failData.code == "1"){//文件已上传过，无需再次上传
                                        layer.close(formData.md5LoadIndex);
                                        formData.md5LoadNoNeedShow = true;
                                        layer.msg('文件已上传过，无需再次上传。', {
                                            icon: 1,
                                            time: 3000
                                        });
                                        uploader.skipFile(file);
                                    }else if(failData.code == "2"){//MD5不一致，新文件上传
                                        formData.uploadedChunksMd5 = undefined;
                                        uploader.option( 'formData', formData);
                                    }
                                    deferred.resolve();
                                }.bind(this));
                        }.bind(this));
                return deferred.promise();
            }.bind(this)
            , beforeSend: function (block) {
                var owner = this.owner;
                var deferred = WebUploader.Base.Deferred();
                uploader.md5File( block.blob )
                    .fail(function() {
                        alert("计算分片md5值失败。");
                        deferred.reject();
                    })
                    .then(function(val) {
                        var remoteMd5 = formData.uploadedChunksMd5;
                        var blockAlreadyUploaded = false;
                        if(remoteMd5){
                            for(var i=0; i<remoteMd5.length; i++){
                                if(remoteMd5[i] == val){
                                    blockAlreadyUploaded = true;
                                    break;
                                }
                            }
                        }
                        if(blockAlreadyUploaded){
                            deferred.reject();//不再上传此分片
                        }else{
                            deferred.resolve();//继续上传此分片
                        }
                    }.bind(this));
                return deferred.promise();

            }
            , afterSendFile: function (file) {
                //alert("afterSendFile");
                layer.close(formData.md5LoadIndex);
            }
        });

        uploader = new WebUploader.Uploader({
            // swf文件路径
            swf: '../plugins/webuploader/Uploader.swf',
            // 文件接收服务端。
            server: uploadUrl,
            // 选择文件的按钮。可选。
            // 内部根据当前运行是创建，可能是input元素，也可能是flash.
            pick: '#atr-picker-'+attachmentId,
            //如果是单一文件上传，设置上传队列大小限制为1。
            fileNumLimit: 1,
            // 不压缩image, 默认如果是jpeg，文件上传前会压缩一把再上传！
            compress: false,
            // 后缀和mime限制
            accept:acceptObj,
            // post参数
            formData:formData,
            // 选中立即上传
            auto:true,
            // 分片，上传断点续传使用
            chunked:chunked,
            chunkRetry:0,
            prepareNextFile: false,
            duplicate: false
        });

        uploader.on( 'uploadStart', function (file) {
            // while(!formData.fileMd5){
            //     //等待md5计算完毕
            //     console.log("waiting for md5");
            // }
            // console.log("got md5:"+formData.fileMd5);
        }.bind(this));

        // 当有文件被添加进队列的时候
        uploader.on( 'beforeFileQueued', function( file ) {
            //文件名超长不允许上传
            if (file.name.length > 32) {
                uploader.removeFile(file);//从上传队列移除
                alert("文件名超长（包括后缀总长度超过32），不允许上传。");
                return false;
            }
            //0byte文件不允许上传
            if (file.size == 0) {
                uploader.removeFile(file);//从上传队列移除
                alert("文件为空，不允许上传。");
                return false;
            }
            //验证size，由于使用了分片，导致
            if (file.size > args.limit * 1024 * 1024 ){
                uploader.removeFile(file);//从上传队列移除
                alert("文件大小超过了"+args.limit+"M，不允许上传。");
                return false;
            }

        }.bind(this));

        // 当有文件被添加进队列的时候
        uploader.on( 'fileQueued', function( file ) {
            //在输入框里显示已加入队列的文件名
            $('#atr-attachInput-' + args.attachmentId).val(
                uploader.getFiles().reduce(
                    function(p,c,i,a){
                        return {name:p.Status=="cancelled"?"":p.name+","+c.Status=="cancelled"?"":c.name}
                    }
                ).name);
        }.bind(this));

        // 文件上传过程中创建进度条实时显示。
        uploader.on( 'uploadProgress', function( file, percentage ) {
            this.addProgress(args,percentage,file);//添加进度条
        }.bind(this));

        //文件成功、失败处理
        uploader.on( 'uploadSuccess', function( file ) {
            uploader.removeFile(file);//从上传队列移除
            this.deleteProgress(args);//删除进度条
            this.showBackDatas(args);//重新回显
            //上传成功后的回调
            if(args.uploadSuc){
                args.uploadSuc(file);
            }
        }.bind(this));

        uploader.on( 'uploadError', function( file ) {
            var originInputVal = $('#atr-attachInput-' + args.attachmentId).val();
            var newInputVal = '上传'+file.name+'出错。';
            //先提示上传出错，3秒后显示原内容。
            $('#atr-attachInput-' + args.attachmentId).val(newInputVal);
            setTimeout(function(){$('#atr-attachInput-' + args.attachmentId).val(originInputVal);}.bind(this),3000);
            this.deleteProgress(args);//删除进度条
            uploader.removeFile(file);//从上传队列移除
        }.bind(this));

        uploader.on( 'uploadComplete', function( file ) {
            this.deleteProgress(args);//删除进度条
            formData.md5LoadNoNeedShow = true;
        }.bind(this));

        uploader.on("error",function (type){
            if (type=="Q_TYPE_DENIED"){
                alert("请上传"+args.suffix+"格式文件");
            }
        });

        this.showBackDatas(args);
    }

    //生成uuid的函数
    guid() {
        return 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'.replace(/[x]/g, function(c) {
            var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
            return v.toString(16);
        });
    }

    //根据data整理出回显用数据
    refreshDatas(args){
        var attachments = args.data.data;
        args.showBack = {};
        args.showBack.valid = true;

        if(attachments.length == 0){
            args.showBack.valid = false;
            return;
        }
        args.showBack.attachmentCount = attachments.length;
        args.showBack.attachments = attachments;
        args.showBack.attachments.map(
            function(item){
                item.attachmentDeletePath=args.attachUrlBase+"/deleteById/"+item.id;
                item.attachmentHttpPath=args.attachUrlBase+"/downloadById/"+item.id;
                item.attachmentInProgress=item.attachmentSize.indexOf("/")>=0;//未上传完毕的附件
                console.log("["+item.attachmentName+"] size:"+item.attachmentSize);
                return item
            }.bind(this)
        );
        //设置回显附件名
        args.showBack.attachmentName = attachments.reduce(function(p,c,i,a){
            return {
                attachmentName:
                        this.getPercent100(p.attachmentSize,p.attachmentName)
                        +"， "
                        +this.getPercent100(c.attachmentSize,c.attachmentName),
                attachmentSize:
                        undefined
            }
        }.bind(this)).attachmentName;
        if(!args.showBack.attachmentName){
            args.showBack.attachmentName="";
        }
        args.showBack.attachmentHttpPath = args.attachUrlBase+"/download/"+args.attachmentId;
        if(attachments.length == 1) {
            args.showBack.id = attachments[0].id;
        }
    }

    showBackDatas(args){
        this.fetchData(args).then(
            function(data){
                args.data = data;
                this.refreshDatas(args);

                var attachmentId = args.attachmentId;
                var showBack = args.showBack;

                //整理回显
                var attacher = $("#atr-main-"+attachmentId);
                if(args.atrType=="default"){
                    $("#atr-attachInput-"+attachmentId).val("");
                    if($("#atr-readonlyInputRow-"+attachmentId)[0]){
                        $("#atr-readonlyInputRow-"+attachmentId).replaceWith(this.getReadonlyInputRow(args));
                    }else{
                        attacher.append(this.getReadonlyInputRow(args));
                    }
                }else if(args.atrType=="aLink"){
                    $("#atr-attachInput-"+attachmentId).val("");
                    if($("#atr-aLinkRow-"+attachmentId)[0]){
                        $("#atr-aLinkRow-"+attachmentId).replaceWith(this.getALinkRow(args));
                    }else{
                        attacher.append(this.getALinkRow(args));
                    }
                }else if(args.atrType=="img42"){
                    $("#atr-attachInput-"+attachmentId).val("");
                    if($("#atr-imgRow-"+attachmentId)[0]){
                        $("#atr-imgRow-"+attachmentId).replaceWith(this.getImgRow(args));
                    }else{
                        attacher.append(this.getImgRow(args));
                    }
                }

                this.registEvents(args);

                this.swiper = new Swiper('.swiper-container', {
                    zoom: true,
                    pagination: '.swiper-pagination',
                    nextButton: '.swiper-button-next',
                    prevButton: '.swiper-button-prev',
                    slidesPerView: 5,
                    spaceBetween: 10,
                    preventClicks : false
                });

            }.bind(this)
        );
    }

    //为生成的新的控件，注册事件
    registEvents(args){
        //aLink及img模式下，一个附件的删除
        $('a[id^="atr-aLinkDeleteBtn-"], a[id^="atr-imgDeleteBtn-"]').off('click').on('click',function(event){
            var id = $(event.target).attr("atrId");
            AjaxRequest.ajax({url:args.attachUrlBase+"/deleteById/"+id}).then(function(deleteResult){
                if(deleteResult.data){
                    this.showBackDatas(args);
                }
            }.bind(this));
        }.bind(this));
        //default模式下，附件整体的删除
        $('div[id^="atr-delBtn-"]').off('click').on('click',function(event){
            var attachmentId = $(event.target).attr("atrAttachmentId");
            AjaxRequest.ajax({url:args.attachUrlBase+"/delete/"+attachmentId}).then(function(deleteResult){
                if(deleteResult.data){
                    this.showBackDatas(args);
                }
            }.bind(this));
        }.bind(this));
    }

    getInputRow(args, type){

        var attachmentId = args.attachmentId;
        var inputRow = $('<div id="atr-inputRow-'+attachmentId+'" class="row"></div>');

        var inputWidth = args.inputWidth?" width:"+args.inputWidth+"px ":"";

        var input = undefined;
        if(type=="default"){
            input = $('<input type="text" id="atr-attachInput-'+attachmentId+'" class="col-sm-7 uploader-input" ' +
            'style="float: left; ' + inputWidth + '" placeholder=" " value="" >');
        }else if(type=="aLink" || type=="img42"){
            input = $('<input type="text" id="atr-attachInput-'+attachmentId+'" class="col-sm-9 uploader-input" ' +
            'style="float: left; ' + inputWidth + '" placeholder=" " value="" >');
        }

        var btns = undefined;
        if(type=="default"){
            btns = $('  <div class="col-sm-5 btns" align="center"> '+
                        '<div id="atr-picker-'+attachmentId+'" class="uploader-picker">选择文件</div> '+
                        '<div id="atr-delBtn-'+attachmentId+'" atrAttachmentId="'+attachmentId+'" class="btn btn-primary">删除</div> '+
                    '</div>');
        }else if(type=="aLink" || type=="img42"){
            btns = $('  <div class="col-sm-3 btns" align="center">\
                        <div id="atr-picker-'+attachmentId+'" class="uploader-picker">选择文件</div>\
                    </div>');
        }

        inputRow.append(input);
        inputRow.append(btns);

        return inputRow;
    }

    getReadonlyInputRow(args){

        var attachmentId = args.attachmentId;
        var showBack = args.showBack;

        if(!showBack.valid){
            return;
        }

        if(showBack.attachmentCount == 1){
            var readonlyInputRow = $('<div id="atr-readonlyInputRow-'+attachmentId+'" class="row"></div>');
            var percentName = this.getPercent100(showBack.attachments[0].attachmentSize, showBack.attachments[0].attachmentName);
            var readonlyInput = $('<input id="atr-readonlyInput-'+attachmentId+'" class="col-sm-12" readonly="readonly" value="'+percentName+'"></input>');
            readonlyInputRow.append(readonlyInput);
            return readonlyInputRow;
        }else{
            var readonlyInputRow = $('<div id="atr-readonlyInputRow-'+attachmentId+'" class="row"></div>');
            var readonlyInputValue = "";
            showBack.attachments.forEach(function(v,i,a){
                readonlyInputValue = readonlyInputValue +"， "+ this.getPercent100(v.attachmentSize, v.attachmentName);
            }.bind(this));
            if(readonlyInputValue.length){
                readonlyInputValue = readonlyInputValue.substr(2);
            }
            var readonlyInput = $('<input id="atr-readonlyInput-'+attachmentId+'" class="col-sm-12" readonly="readonly" value="'+readonlyInputValue+'"></input>');
            readonlyInputRow.append(readonlyInput);
            return readonlyInputRow;
        }
    }

    getALinkRow(args){

        var attachmentId = args.attachmentId;
        var showBack = args.showBack;

        if(!showBack.valid){
            return;
        }

        if(showBack.attachmentCount == 1){
            var aLinkRow = $('<div id="atr-aLinkRow-'+attachmentId+'" class="row"></div>');
            var linkAddr = showBack.attachments[0].attachmentInProgress?"":args.attachUrlBase+"/downloadById/";
            var percentName = this.getPercent100(showBack.attachments[0].attachmentSize, showBack.attachmentName);
            var aLinkAndDeleteBtn = this.getALinkAndDeleteBtn(percentName, linkAddr, showBack.id);
            aLinkRow.append(aLinkAndDeleteBtn[0]);
            if(!args.isShowBack)
                aLinkRow.append(aLinkAndDeleteBtn[1]);
            return aLinkRow;
        }else{
            var aLinkRow = $('<div id="atr-aLinkRow-'+attachmentId+'" ></div>');
            showBack.attachments.forEach(function(v,i,a){
                var linkAddr = v.attachmentInProgress?"":args.attachUrlBase+"/downloadById/";
                var percentName = this.getPercent100(v.attachmentSize, v.attachmentName);
                var aLinkRowSub = $('<div id="atr-aLinkRow-sub-'+v.id+'" class="row"></div>');
                var aLinkAndDeleteBtn = this.getALinkAndDeleteBtn(percentName,linkAddr, v.id);

                aLinkRowSub.append(aLinkAndDeleteBtn[0]);
                if(!args.isShowBack)
                    aLinkRowSub.append(aLinkAndDeleteBtn[1]);
                aLinkRow.append(aLinkRowSub);
            }.bind(this));
            return aLinkRow;
        }
    }

    getALinkAndDeleteBtn(name, pathSuffix, id){

        var name = name ? name : "无";
        var downloadPath = pathSuffix+id;
        var hrefStr = 'href='+downloadPath+' target="_blank" ';
        //生成回显用附件超链接
        var aLink = $('<span id="atr-aLinkSpan-'+id+'" class="col-sm-9">' +
        '<a id="udr-aLink-'+id+'" '+hrefStr+'>'+name+'</a></span>');
        //生成删除按钮
        var aLinkDeleteBtn = $('<div id="atr-aLinkDeleteDiv-'+id+'" class="col-sm-3" align="center" >' +
        '<span id="atr-aLinkDeleteSpan-'+id+'" style="height:100%" >' +
        '<a class="uploaderDelete" style="cursor:pointer;"' +
        ' id="atr-aLinkDeleteBtn-'+id+'" atrId="'+id+'"/></span></div>');
        return [aLink, aLinkDeleteBtn];
    }

    getImgRow(args){

        var attachmentId = args.attachmentId;
        var showBack = args.showBack;

        if(!showBack.valid){
            return;
        }

        var imgRow = $('<div id="atr-imgRow-'+attachmentId+'" class="row swiper-container"></div>');
        var imgWrapper = $('<div class="swiper-wrapper">');
        var naviPrev = $('<div class="swiper-button-prev"></div>');
        var naviNext = $('<div class="swiper-button-next"></div>');

        if(showBack.attachmentCount == 1){
            var imgSlide = $('<div class="swiper-slide">');
            var imgAndDeleteBtn = this.getImgAndDeleteBtn(showBack.attachmentHttpPath, showBack.id);
            imgSlide.append(imgAndDeleteBtn[0]);
            imgSlide.append(imgAndDeleteBtn[1]);
            imgWrapper.append(imgAndDeleteBtn);
            imgRow.append(imgWrapper);
            imgRow.append(naviPrev);
            imgRow.append(naviNext);
            return imgRow;
        }else{
            showBack.attachments.forEach(function(v,i,a){
                var imgSlide = $('<div class="swiper-slide">');
                var imgAndDeleteBtn = this.getImgAndDeleteBtn(v.attachmentHttpPath, v.id);
                imgSlide.append(imgAndDeleteBtn[0]);
                imgSlide.append(imgAndDeleteBtn[1]);
                imgWrapper.append(imgSlide);
            }.bind(this));
            imgRow.append(imgWrapper);
            imgRow.append(naviPrev);
            imgRow.append(naviNext);
            return imgRow;
        }
    }

    getImgAndDeleteBtn (path, id){

        var img = $('<a href="'+path+'"><img src="'+path+'?rand='+this.guid()+'" style="width:100px"></a>');
        var del = $('<div id="atr-imgDeleteDiv-'+id+'" align="center" >' +
        '<span id="atr-imgDeleteSpan-'+id+'" >' +
        '<a class="uploaderDelete" style="cursor:pointer;"' +
        ' id="atr-imgDeleteBtn-'+id+'" atrId="'+id+'"/></span></div>');

        return [img,del];
    }

    addProgress(args, percentage, file){
        var $percent = $("#atr-main-"+args.attachmentId).find('.progress .progress-bar');
        // 避免重复创建
        if ( !$percent.length ) {
            $percent = $('<div class="row progressRow">' +
                '<div class="col-sm-9 progress progress-striped active" style="padding:0px">' +
                    '<div class="progress-bar" role="progressbar" style="width:0.1%">' +
                    '</div>' +
                '</div>' +
                '</div>').appendTo("#atr-main-"+args.attachmentId).find(".progressRow");

            //不断查询进度，并更新progress
            args.percentIntervalId = window.setInterval(function(){
                    this.updateProgress(args,file);
                }.bind(this)
                ,3000);
        }
    }
    updateProgress(args,file){
        var percentage = "";
        var getData = {};
        getData.attachmentId = args.attachmentId;
        getData.fileName = file.name;
        $.ajax({
            type:"post",
            url:"/attach/get",
            data:getData,
            async:false,
            success:function (result) {
                percentage = result.data.attachmentSize;
            }.bind(this)
        });
        console.log("percentage:"+percentage);

        var pct = 100;
        var inProcess = false;
        if(percentage.indexOf("/")>=0){
            inProcess = true;
            var pctChild = parseInt(percentage.substring(0,percentage.indexOf("/")));
            var pctParent = parseInt(percentage.substr(percentage.indexOf("/")+1));
            pct = (pctChild/pctParent)*100;
        }

        var $percent = $("#atr-main-"+args.attachmentId).find('.progress .progress-bar');
        if ( $percent.length ) {
            $percent.attr('style',"width:"+pct+"%");
        }else{
            inProcess = false;
        }

        if (!inProcess){
            window.clearInterval(args.percentIntervalId);
        }
    }
    deleteProgress(args){
        var $progress = $("#atr-main-"+args.attachmentId).find('.progressRow');
        if($progress.length){
            $progress.fadeOut(1000,function(){$progress.remove();}.bind(this));
        }
    }
    // //获取md5，使用原生html5的fileApi和spark-md5，不使用web-uploader的md5File
    // getMd5(file){
    //         //注意此方法引用了SparkMD5库 library:https://github.com/satazor/SparkMD5
    //
    //         //声明必要的变量
    //         var fileReader = new FileReader(), box = document.getElementById('box');
    //         //文件分割方法（注意兼容性）
    //         var blobSlice = File.prototype.mozSlice || File.prototype.webkitSlice || File.prototype.slice,
    //             //file = document.getElementById("file").files[0],
    //             //文件每块分割2M，计算分割详情
    //             chunkSize = 2097152,
    //             chunks = Math.ceil(file.size / chunkSize),
    //             currentChunk = 0,
    //             //创建md5对象（基于SparkMD5）
    //             spark = new SparkMD5();
    //
    //         //每块文件读取完毕之后的处理
    //         fileReader.onload = function(e) {
    //             console.log("读取文件", currentChunk + 1, "/", chunks);
    //             //每块交由sparkMD5进行计算
    //             spark.appendBinary(e.target.result);
    //             currentChunk++;
    //
    //             //如果文件处理完成计算MD5，如果还有分片继续处理
    //             if (currentChunk < chunks) {
    //                 loadNext();
    //             } else {
    //                 console.log("finished loading");
    //                 //box.innerText = 'MD5 hash:' + spark.end();
    //                 console.info("计算的Hash", spark.end());
    //                 return spark.end();
    //             }
    //         };
    //
    //         //处理单片文件的上传
    //         function loadNext() {
    //             var start = currentChunk * chunkSize, end = start + chunkSize >= file.size ? file.size : start + chunkSize;
    //             fileReader.readAsBinaryString(blobSlice.call(file, start, end));
    //         }
    //
    //         loadNext();
    // }


    //根据参数分数取得其对应的百分比形式
    getPercent100(percent, oldName){

        if(!percent) {
            return oldName
        }
        if(percent.indexOf("/")<0){
            return oldName;
        }
        var percentChild = parseInt(percent.substring(0, percent.indexOf("/")));
        var percentParent = parseInt(percent.substr(percent.indexOf("/")+1));
        var pct = ""+((percentChild/percentParent)*100);
        var pctPoint2 = pct.substring(0,pct.indexOf(".")+3);
        return "(上传进度："+pctPoint2+"%)"+oldName;
    }
    getMime(keys){
        var mimeTbl = this.mimeTbl;
        var keyArr = [];
        if(keys.indexOf(",")>0){
            keyArr = keys.split(",");
        }else{
            keyArr[0] = keys;
        }

        var mimeArr = new Set();
        var index = 0;

        var myDate = new Date();
        console.log("获取mime开始："+myDate.getFullYear()+"-"+myDate.getMonth()+"-"+myDate.getDate()+" "+myDate.getHours()+":"+myDate.getMinutes()+":"+myDate.getSeconds()+":"+myDate.getMilliseconds()); //获取完整的年份(4位,1970-????)

        for(var k=0; k<keyArr.length; k++){
            if( mimeTbl[keyArr[k]] ){
                mimeArr.add(mimeTbl[keyArr[k]]);
            }
        }

        var myDate = new Date();
        console.log("获取mime结束："+myDate.getFullYear()+"-"+myDate.getMonth()+"-"+myDate.getDate()+" "+myDate.getHours()+":"+myDate.getMinutes()+":"+myDate.getSeconds()+":"+myDate.getMilliseconds()); //获取完整的年份(4位,1970-????)

        var mimeResult = "";
        if(mimeArr.size>0){
            mimeArr.forEach(function(value,key){if(mimeResult == ""){mimeResult += value;}else{mimeResult+=","+value;}});
        }
        return mimeResult;
    }

}
