// ezdev-system/common/layout/navbars/user/UserNav.js
import BaseNav from "ezdev-pcview/layout/navbar/BaseNav";
import config from "ezdev-pcview/Config";
import EAttach from "./EAttach";

let layerId;
let isUserImgExist;

export default class extends BaseNav {
    constructor(options) {
        super(Object.assign({
            name: "nav/user",
            order: 20
        }, options));

        this.addEvent("submited", ()=> {
            layer.close(this.layerId);
        })
    }

    resolveTpl(tpl) {
        return require("./tpl/" + tpl + ".html");
    }

    /**
     * 生成html
     */
    generateHtml() {
        return this.resolveTpl("index")(config.user);
    }

    disposeHtml() {
        this.ajaxResource({
            url:'attach/list/'+config.user.id ,
            type: 'post'
        }).then((data) =>{
            isUserImgExist = data.data.length;
            if(isUserImgExist>0){
                $('.userimg').attr('src',"/attach/download/"+config.user.id+'?ran='+Math.random());
            }
        });
    }


    menuDialog(args) {
        $("#user-menu").removeClass("open");
        this.layerId = layer.open({
            type: 1,
            move: args.move,
            title: args.title || '信息',
            area: [args.width || '800px', args.height || '600px'] //宽高
        });

        args.contentId = $('#layui-layer' + this.layerId).find('.layui-layer-content');

        this.render(args).then(function () {
            if($("#userId").length>0){
                var attachid=$("#userId").val();
                if(isUserImgExist==0){
                    $('.userimg').attr('src',"default.jpg");
                }else{
                    $('.userimg').attr('src',"/attach/download/"+attachid+'?ran='+Math.random());
                }
                var atcArgs = {
                    el: '#myeattach'
                    , domopt: "replace"
                    , atrIdSaver: "usBizId"
                    , atrType: 'userImg'
                    , attachmentId: attachid
                    , suffix:"jpg,jpeg,png,bmp"
                    , limit:'3'
                    , single: 'true'
                    , uploadSuc: function (file) {
                        $('.userimg').attr('src',"/attach/download/"+attachid+'?ran='+Math.random());
                        isUserImgExist==1;
                    }
                };
                let attach = new EAttach();
                attach.reWriteHtml(atcArgs);
            }

        });
    }

    cancel() {
        layer.close(this.layerId);
    }

};
