// ezdev-pcview/components/popu/treelist/EpoptreelistHandler.js
/**
 * @描述：事件类，继承PopHandler
 * Created by wangchen on 2017/8/7.
 * PopTreelist对应的Handler，用于事件
 * 树的生成及设置
 */
import PopHandler from "../EPopuHandler";
import HtmlUtils from "../../../utils/HtmlUtils";
import util from "../../utils/util";
import  "ztree/css/zTreeStyle/zTreeStyle.css";
import "ztree";

export default class PopTreelistHandler extends PopHandler{

	constructor(options){
        super(options);
    }

    resolveTpl(tpl) {
        return require("./tpl/" + tpl + ".html");
    }

    /**
     * Index页面组织树初始化
     */
    initTree(args) {
        // this.options1();
        $.ajax({
            cache: false,
            type: "POST",
            url: args.treeUrl,
            dataType: "json",
            success: function (data) {
                $.fn.zTree.init($('#' + args.treeId), this.ztreeSetting(args), data.data);
                var zTree = $.fn.zTree.getZTreeObj(args.treeId);
                var root = zTree.getNodeByParam("id", "root");
                zTree.selectNode(root, false);
                //zTree.expandAll(true);
                zTree.expandNode(root, true, false, false);  //默认展开第一级
            }.bind(this)
        });
    }

    /**
     * 组织机构树配置
     */
    ztreeSetting(args) {
        // this.options1();
        return {
            check: {
                enable: false
            },
            view: {
                selectedMulti: false
            },
            data: {
                simpleData: {
                    enable: true
                }
            },
            callback: {
                onClick: function (event, treeId, treeNode) {
                    util.setClickNodeScrollPos(treeId,event);//设置滚动条到

                    let $form = $(event.target).parents('form.popupTreeListForm');
                    let treeListPara = args.treeListPara;
                    if(typeof(treeListPara)!='undefined' && treeListPara.length > 0){
                        for (let v of treeListPara) {
                            $form.find("#"+v['listid']).val(treeNode[v["tree"]]);
                        }
                    }
                    let $sbtn = $form.find('#treelistSearchBtn[operation="search"]');
                    let searchArgs = HtmlUtils.getArgsByItem($sbtn);
                    searchArgs.form = $form;

                    this.search(searchArgs);
                }.bind(this)
            },
            view: {
                fontCss: function (treeId, treeNode) {
                    return (!!treeNode.highlight) ? {color: "blue", "font-weight": "bold"} : {
                        color: "#000",
                        "font-weight": "normal"
                    };
                }
            }
        }
    }
}
