// ezdev-pcview/components/utils/multiple/Multiple.js
/***
 * @描述：工具类，用于生成多项选择控件的显示区域及提交表单的隐藏域
 * @author:create by wangchen 2017.8.21
 * @用途：复用代码，1、显示区域和隐藏域的html 2、选择的事件
 */
export default class Multiple {

    //生成显示区域及对应隐藏域的html
    generateHtml(id,name) {
        let data = {
            id: id,
            name:name
        };
        return this.getBtnTl()(data || {});
    }

    getBtnTl() {
        return this.resolveTpl("index");
    }

    resolveTpl(tpl) {
        return require("./tpl/" + tpl + ".html");
    }

    //添加一个值到显示区和隐藏域
    appendBackDisplay(id, txt, val) {
        let $dis = $('#' + id + '_dis');
        let $disTxt = $('#' + id + '_dis_txt');
        let $hid = $('#' + id + '_hid');
        let hidval = $hid.val();
        let matchval = ','+hidval+',';
        if (matchval.indexOf(','+val+',') > -1) {
            return; //如果已经存在则不重复添加
        }
        if (hidval !='') {
            hidval += ',' + val;
        } else {
            hidval = val;
        }
        $hid.val(hidval);
        $disTxt.val(hidval);
        $dis.append(`<li class="select2-selection__choice" title="${txt}">
                    <span class="select2-selection__choice__remove" hidval="${val}" role="presentation">×</span>${txt}</li>
                    `);

        this.addBackDisplayRemoveEvent($dis, $hid,$disTxt);
    }

    //删除一个值
    removeBackDisplay(id, val) {
        // 2018-05-24 yangfei 增加删除值所在的显示区限制
        let $dis = $('#' + id + '_dis');
        $dis.find(`span[hidval="${val}"].select2-selection__choice__remove`).parent().remove();
        let $hid = $('#' + id + '_hid');
        let $disTxt = $('#' + id + '_dis_txt');
        let hidval = $hid.val();
        if(hidval=='' || hidval == null){
            return;
        }
        let matchval = hidval.split(',');
        let arrs = matchval.filter(item => item !== val);
        $hid.val(arrs.join(','));
        $disTxt.val(arrs.join(','));

    }

    //设置页面控件回显的值
    setBackDisplay(id, txts, vals) {
        //$('#' + id + '_dis').val(txts.join(','));
        let html = [];
        if (txts && txts.length > 0) {
            let val = null;
            for (let i = 0; i < txts.length; i++) {
                if (i == 0) {
                    val = vals[i];
                } else {
                    val = ',' + vals[i];
                }
                html.push(`<li class="select2-selection__choice" title="${txts[i]}">
                    <span class="select2-selection__choice__remove" hidval="${val}" role="presentation">×</span>${txts[i]}</li>
                    `);
            }

        }
        let $dis = $('#' + id + '_dis');
        let $disTxt = $('#' + id + '_dis_txt');
        $dis.html(html.join(''));
        let $hid = $('#' + id + '_hid');
        $hid.val(vals.join(','));
        $disTxt.val(vals.join(','));
        this.addBackDisplayRemoveEvent($dis, $hid,$disTxt);

        var allTxt = "";
        $dis.find('li.select2-selection__choice').each(function (index, item) {
            if(allTxt == ""){
                allTxt = $(item).attr('title');
            }else {
                allTxt = allTxt + ',' + $(item).attr('title');
            }
        });
        $dis.removeAttr('title').attr('title',allTxt);
    }

    setValidform(id,datatype){
        if(typeof(datatype)=='undefined' || datatype == null){
            return;
        }
        let $disTxt = $('#' + id + '_dis_txt');
        $disTxt.attr('datatype',datatype);
    }

    //设置显示区域叉的删除事件
    addBackDisplayRemoveEvent($dis, $hid,$disTxt) {
        $dis.find('span[hidval].select2-selection__choice__remove').each(function() {
            let $this = $(this);
            $this.off('click');
            $this.on('click', function() {
                let $dis = $this.parent().parent();

                $this.parent().remove();
                let val = $hid.val();
                val = val.replace($(this).attr('hidval'), '');
                $hid.val(val);
                $disTxt.val(val);

                var allTxt = "";
                $dis.find('li.select2-selection__choice').each(function (index, item) {
                    if(allTxt == ""){
                        allTxt = $(item).attr('title');
                    }else {
                        allTxt = allTxt + ',' + $(item).attr('title');
                    }
                });
                $dis.removeAttr('title').attr('title',allTxt);
            });
        });

    }



}
