// ezdev-system/business/authorize/roleuser/Handler.js
/**
 * Created by suntf on 2017/7/24
 */
import Handler from "ezdev-pcview/handler/Handler";

export default new Handler({
    resolveTpl: function (tpl) {

        return require("./tpl/" + tpl + ".html");
    },
    initHandler:function(){
        this.options = {
            leftId : 'leftcontent',
            rightId : 'rightcontent',
            roleListId:'roleList',
            rightTreeId:'tree_id',
            rightTreeUrl:'system/core/user/orgusertree',
            pageId : 'pager',
            selectedRoleId : '',
            selectedRescId : '',
            lastIds : '',
            selectedRoleObj:{}
        };
    },

    initRoleList:function(){
        let args={
            contentId : "list",
            data:"search_LIKE_name=&page=1",
            form:"searchForm",
            tpl :"list",
            url :"/system/core/role/list"
        };
        $('.pagination2').jqPagination({
            current_page: 1, //设置当前页 默认为1
            max_page	: 1,
            page_string : '第{current_page}页,共{max_page}页',
            paged		: function(page) {
                $("#" + this.options.leftId+' #page').val(page);
                this.searchNew(args).then(function(){
                    $("#example2 tr:first").click();
                });
            }.bind(this)
        });
        //this.searchRoleList();
    },
    /*新的查询方法*/
    searchNew:function(args){
        args.form = args.form || 'searchForm';
        if (args.url) {
            args.contentId = args.contentId||'list';
            let form = $(`#${args.form}`);
            if (form.length > 0) {
                args.data = form.serialize();
                args.tpl = args.tpl||'list';
                return this.render(args).then(function (data) {
                    let maxpage=parseInt(data.data.totalPages);
                    let curpage=parseInt(data.data.pageNum);
                    $(".pagination2").jqPagination("option", {
                        max_page : maxpage,
                        current_page :curpage,
                        trigger : false
                    });
                }.bind(this));
            } else {
                throw new Error("参数无效，DOM中无法找到id为[" + args.form + "]的Form！");
            }

        } else {
            throw new Error("参数无效，请传递如{url:/demo/mybatisuser/pages-(必选),type:post-(默认),tpl:list-(默认),contentId:list-(默认)}的JS对象");
        }

    },
    /*新的查询方法ed*/
    onSearchNewed:function(args){
        this.initUserTree(args);
        $("#example2 tr:first").click();
    },
    /**
     * 点击角色，取得授权用户，更新用户tree
     */
    initUserTree : function(args) {
        this.dialog = null;
        var tree_setting = {
            check : {
                enable : true,
                chkboxType: { "Y" : "ps", "N" : "ps" }
            },
            data : {
                simpleData : {
                    enable : true
                }
            },
            view :{
                fontCss: function(treeId, treeNode){
                    return (!!treeNode.highlight) ? {color:"blue", "font-weight":"bold"} : {color:"#000", "font-weight":"normal"};
                }
            }
        };
        $.ajax({
            cache : false,
            async : false,// 同步加载
            type : "POST",
            url : this.options.rightTreeUrl,
            dataType : "json",
            success : function(data) {
                var nodes=data.data;
                $.each(nodes,function(i,item){
                    if(item.type=='user'){
                        item.iconSkin="user";
                    }
                });
                $.fn.zTree.init($(`#${this.options.rightTreeId}`),tree_setting, nodes);
                var zTree = $.fn.zTree.getZTreeObj(`${this.options.rightTreeId}`);
                var root = zTree.getNodesByFilter(function (node) { return node.level == 0 }, true);
                zTree.expandNode(root, true, false, false);  //默认展开第一级
            }.bind(this)
        });
    },
    /**
     * 点击角色，取得授权用户，更新用户tree
     */
    selRole:function(args){
        this.options.selectedRoleObj=args;
        $.fn.zTree.destroy("dataRuleTree");
        var roleid=args.objid;
        var rolename=args.objname;
        //点击过后的tr变成蓝色，其他tr恢复原色
        $(`#${roleid}`).addClass("active").siblings().removeClass('active');
        if(!roleid){
            $("#"+this.options.rightId+" #btnSure").attr('disabled',true);
        }else{
            $("#"+this.options.rightId+" #btnSure").attr('disabled',false);
        }
        $("#"+this.options.rightId+" #hidRoleid").val(roleid);
        $("#"+this.options.rightId+" #spnCurRole").text(rolename);
        if(roleid){
            var url = "system/core/role/usersbyroleid/"+roleid;
            $.ajax({
                cache : false,
                async : false,// 同步加载
                type : "GET",
                url : url,
                dataType : "json",
                success : function(data) {
                    var zTree = $.fn.zTree.getZTreeObj(this.options.rightTreeId);
                    zTree.checkAllNodes(false); //用户数选择状态初始化
                    if(data.data.length>0){
                        data.data.forEach(function (item, i){
                            var node = zTree.getNodeByParam("id",item.id);
                            if(node){
                                node.checked=true;
                                //zTree.updateNode(node,true);//父节点级联选中
                                zTree.updateNode(node);
                                zTree.expandNode(node.getParentNode(), true, false, false);  //展开选中节点
                            }

                        });
                    }
                }.bind(this)
            });
        }
    },
    bindUsers:function(){
        //选中角色
        var roleid= $("#"+this.options.rightId+" #hidRoleid").val();
        //选中用户
        var treeObj = $.fn.zTree.getZTreeObj(this.options.rightTreeId);
        var chknodes = treeObj.getCheckedNodes(true);
        var nodes="";
        chknodes.forEach(function(item, i) {
            if(item.type=='user'){
                nodes+= item.id+",";
            }
        });
        if(roleid.length==0 ){
            //TODO:替换为标准提示框
            layer.msg(`请选择绑定的角色`);
            return;
        }
        var param = {"roleid":roleid,"userids":nodes};
        var url="system/core/authorize/bindroleusers";
        $.ajax({
            cache : false,
            data:param,
            type : "POST",
            url : url,
            dataType : "json",
            success : function(data) {
            }.bind(this)
        });
    },
    bindReset:function(){
        var element = this.options.selectedRoleObj;
        this.selRole(element);
    }
});
