// ezdev-pcview/utils/MiscUtils.js
/**
 * Created by zhaofs on 2017/5/27.
 * --------------------
 * - 汇合工具 -
 * --------------------
 * 各种常用工具大结合。
 */

//add lhw 20160809
export default {
    instanceHash: {},
    /**
     * 获取唯一实例
     * @param className
     */
    getInstance: function (className) {
        if (this.instanceHash[className]) {
            return this.instanceHash[className];
        } else {
            return undefined;
        }
    },
    /**
     * 把实例加入hash表
     * @param className
     * @param instance
     */
    setInstance: function (className, instance) {
        this.instanceHash[className] = instance;
    },
    //add end


    //ADD LHW 20151126 追加获取uuid函数，用于生成唯一的id。 START
    guid: function () {
        return 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'.replace(/[x]/g, function (c) {
            let r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    },
    //ADD LHW 20151126 END
    initdate: function (args) {
        if (!args || !args.length) {
            return;
        }
        for (let i = 0; i < args.length; i++) {
            let id = args[i];
            $("#" + id).datepicker({format: 'yyyy-mm-dd', autoclose: true, language: 'zh-CN', todayHighlight: true});
            $("#" + id).attr("readonly", "readonly");
            $("#" + id).css("background-color", "#ffffff");
            $("#" + id).css("cursor", "pointer");
        }
    },
    //add lhw 20170217 模块片段http接口调用功能（工作流中的表单使用）
    analyzeWindowHref: function () {

        let windowHref = {};
        //原始href：localhost/#FullScreen/tpl/demo/message/url/fcc/sqb/c?id=xxxxx
        //转换为：FullScreen/tpl/demo/message/url/fcc/sqb/c?id=xxxxx
        windowHref.href = this.getAnchorUrl();

        //如果href为全屏展示模块片段，向util中放入所有href信息。
        if (windowHref.href && windowHref.href.indexOf("FullScreen") === 0) {
            windowHref.isFullScreen = true;

            let tplAndUrl = windowHref.href.substr(11);//形如：tpl/demo/message/url/fcc/sqb/c?id=xxxxx
            let urlIndex = tplAndUrl.indexOf("/url/");
            if (urlIndex > 0) {
                windowHref.fsHasUrl = true;
                windowHref.fsUrl = tplAndUrl.substr(urlIndex + 5);
                windowHref.fsTpl = tplAndUrl.substring(4, urlIndex);
            } else {
                windowHref.fsHasUrl = false;
                windowHref.fsTpl = tplAndUrl.substr(4);
            }
        } else {
            windowHref.isFullScreen = false;
        }

        this.setInstance("windowHref", windowHref);
    },
    getWindowHref: function () {
        return this.getInstance("windowHref");
    },
    //add end

    /**
     * 锚点刷新
     */
    refreshHash: function () {
        window.ezdev.dispatcher.dispatch();
    },


    //add lhw 20180530 按钮提取到左上角-list选中项目的id的拼接 start
    /**
     * 根据args.url判断是否需要提取list中选中项的id添加到url末尾
     * 根据参数toggle状态（'on'或'off'）设置页面上的单选/多选toggle的状态，如果页面上有toggle的话
     * （'on'为单选，'off'为多选）
     */
    checkItems: function(id, args, countThreshold, toggle) {

        var slashJudge = false;
        if( countThreshold > 0 ){//限制item个数，因此是修改/查看/删除，需限定url的末尾为slash（追加格式）
            slashJudge = true;
        }

        if( slashJudge ) {
            if (!args.url || (args.url.substr(-1)!="/" && args.url.substr(-1)!="\\") ){
                return true;//url末尾不是slash，说明无需抽取checkbox的选中id，直接返回.
            }
        }

        //所有选中状态的checkbox
        var checkedIds = $('#list input.id:checked');

        //1，检查选中数量
        if ( !(checkedIds.length >= countThreshold) ) {//选中的id数目不大于阈值时
            //设置goBackAfterFunc，一旦check不通过需要goBack时，执行此函数
            layer.msg("请选择至少 "+countThreshold+" 条记录");
            // //回退到原页面时的处理
            // var resetToggleFunc = function () {
            //     if( $('.sm-toggle').length > 0 ){//页面存在“单选/多选”toggle
            //         $('.sm-toggle').bootstrapToggle(toggle);//设置为参数状态
            //     }
            // };
            // this.setInstance(id+"-listAfterFunc", resetToggleFunc);
            return false;
        }

        //2，根据单选/多选状态把选中的id收集到args.urlSuffix（用于）
        var checkedIdArray = new Array();
        if( toggle == 'on' ) {//单选的话，关闭除第一个之外的checkbox
            checkedIds.each(function (index, item) {
                if( index == 0 ){
                    checkedIdArray.push($(item).val());
                }else{
                    $(item).removeAttr("checked");
                }
            });
        }else if( toggle == 'off' ){//多选的话，记录所有选中id
            checkedIds.each(function (index, item) {
                checkedIdArray.push($(item).val());
            });
        }
        args.urlSuffix = checkedIdArray;

        var checkAllState = $(".idAll").prop('checked');

        //3，回退到原页面时的处理
        var resetToggleListFunc = function () {
            if( $('.sm-toggle').length > 0 ){//页面存在“单选/多选”toggle
                $('.sm-toggle').bootstrapToggle(toggle);//设置为参数状态
            }
            $(".idAll").prop('checked', checkAllState);
            checkedIdArray.forEach(function (item, index) {
                $('#list input.id[value="'+item+'"]').prop("checked",true);;
            });
        };
        this.setInstance(id+"-listAfterFunc", resetToggleListFunc);

        return true;
    },

    /**
     * 所有totle按钮的初始化
     */
    initToggles: function () {
        $('.sm-toggle').bootstrapToggle();
        $('.sm-toggle').change(function() {
            var checked = $(this).prop('checked');
            if( checked ){//单选
                //注册click事件-单选
                $('#list .id').each(function(){
                    $(this).click(function(){
                        if($(this).prop('checked')){
                            $('#list .id').prop('checked',false);
                            $(this).prop('checked',true);
                        }
                    });
                });
                $('#list .id:checked').each(function(index, item){
                    if( index != 0 ){
                        $(item).prop('checked',false);
                    }
                });
            }else{
                //注册click事件-多选
                $('#list .id').each(function(){
                    $(this).off("click");
                });
            }
        });
        $('.sm-toggle').bootstrapToggle('on');
    },

    /**
     * 初始化全选按钮
     */
    initCheckAll: function () {
        //全选,设置chheckbox name='all' tbody id=tb
        $(".idAll").click(function () {
            if (this.checked) {
                if( $('.sm-toggle').prop('checked') ){//单选状态
                    $("#list .id:first").prop("checked", true);
                }else {
                    $("#list .id").prop("checked", true);
                }
            } else {
                $("#list .id").prop("checked", false);
            }
        });
    }
    //add end
};
