// ezdev-pcview/components/view/single/EPanelViewSingle.js
/**
 * Created by wangchen on 2018/3/31.
 * 还缺少事件的支持
 */
import BaseComponent from "../../BaseComponent";

'use strict';
let __instance = null;
export default class EPanelView extends BaseComponent{
    constructor() {
        super();
        this.name = 'e-panelviewsingle';
        this.handler = null;
    }

    static instance() {
        if (__instance == null) {
            __instance = new EPanelView();
        }
        return __instance;
    }

    /**
     * @Override
     */
   /* setArgsBeforeRw(args, options) {
        args = super.setArgsBeforeRw(args, options);
        args.disgroup = args.disgroup || [];//显示分组的顺序设置，如果没有则不按照分组分片显示

    }*/

    /**
     * @Override
     */
    generateHtml(args) {
        this.adaptableData(args);
        return this.resolveTpl("panelviewsingle")(args || {});
    }

    /**
     * 适配数据，适配后台的数据格式符合组件显示用的格式
     * 将每组数据按照分组给组合起来
     * @param args
     */
    adaptableData(args){
        let disgroup = args.disgroup;
        let data = args.data;
        let databygroup = {};// 每个组属性带有list数据
        if(disgroup.length > 0 ){

            for(let obj of data){
                if(typeof (databygroup[obj.group]) == 'undefined' ||  !(databygroup[obj.group] instanceof  Array)){
                    databygroup[obj.group] = [];
                }
                databygroup[obj.group].push(obj);
            }
        }else{
            disgroup.push('none');
            databygroup['none'] = data;
        }

        args.databygroup = databygroup;
    }

    resolveTpl(tpl) {
        return require("./tpl/" + tpl + ".html");
    }

}
