// ezdev-system/business/authorize/rescuser/Handler.js
/**
 * Created by zhaofs on 2017/5/6.
 * 演示Handler应用
 */
import Handler from "ezdev-pcview/handler/Handler";

export default new Handler({
    options1: function () {
        this.options = {
            leftId: 'leftcontent',
            rightId: 'rightcontent',
            leftTreeId: 'rescTree',
            leftTreeUrl: 'system/core/resource/list',
            rightTreeId:'orguserTree',
            rightTreeid:'orguserTree',
            rightTreeUrl:'system/core/user/orgusertree',
            lastIds: ''
        };
        return this;
    },

    resolveTpl: function (tpl) {
        return require("./tpl/" + tpl + ".html");
    },

    onIndexed: function () {
        this.options1();
        this.initRescTree();
        this.initUserTree();

    },


    /**
     * 资源树
     */
    initRescTree : function(args) {

        var tree_setting = {
            check : {
                enable : false
               /* chkboxType: { "Y" : "ps", "N" : "ps" }*/
            },
            data : {
                simpleData : {
                    enable : true
                }
            },
            callback : {
                onClick : function (event, treeId, treeNode){
                    //if(treeNode.type=='resource'){
                    this.selResc(treeNode.id,treeNode.name);
                    //}
                }.bind(this),
                beforeClick: function(treeId, treeNode){
//								if(treeNode.type=="bizsys"){
//									return false;
//								}
                    if(treeNode.type=="root"){
                        return false;
                    }
                }.bind(this)
            },
            view :{
                fontCss: function(treeId, treeNode){
                    return (!!treeNode.highlight) ? {color:"blue", "font-weight":"bold"} : {color:"#000", "font-weight":"normal"};
                }
            }
        };
        $.ajax({
            cache : false,
            async : false,// 同步加载
            type : "POST",
            url : this.options.leftTreeUrl,
            dataType : "json",
            success : function(data) {
                $.fn.zTree.init($('#' + this.options.leftTreeId),tree_setting, data.data);
                var zTree = $.fn.zTree.getZTreeObj(this.options.leftTreeId);
                //zTree.expandAll(true);
                if(data.data.length == 0){
                    $("#"+this.options.rightId+" #btnSure").attr('disabled',true);
                }else{
                    $("#"+this.options.rightId+" #btnSure").attr('disabled',false);
                }
            }.bind(this)
        });
    },
    /**
     * 初始化用户树tree
     */
    initUserTree: function () {
        var tree_setting = {
            check: {
                enable: true,
                chkboxType: { "Y" : "ps", "N" : "ps" }
            },
            data: {
                simpleData: {
                    enable: true
                }
            }
        };
        $.ajax({
            cache : false,
            async : false,// 同步加载
            type : "POST",
            url : this.options.rightTreeUrl,
            dataType : "json",
            success : function(data) {
                var nodes=data.data;
                $.each(nodes,function(i,item){
                    if(item.type=='user'){
                        item.iconSkin="user";
                    }
//						if(item.type=='org'){
//							item.nocheck=true;
//						}
                });
                $.fn.zTree.init($('#' + this.options.rightTreeId),tree_setting, nodes);
                var zTree = $.fn.zTree.getZTreeObj(this.options.rightTreeId);
                //zTree.expandAll(true);//全部展开
                var root = zTree.getNodesByFilter(function (node) { return node.level == 0 }, true);
                zTree.expandNode(root, true, false, false);  //默认展开第一级

                //默认选中第一个用户
                var rescTree = $.fn.zTree.getZTreeObj(this.options.leftTreeId);
                var nodes = rescTree.getNodesByParam("type", "bizsys");
                rescTree.selectNode(nodes[0], true);
                rescTree.expandNode(nodes[0].getParentNode(), true, false, false);
                //显示用户的绑定角色
                this.selResc(nodes[0].id, nodes[0].name);
            }.bind(this)
        });
    },
    /**
     * 获取全部角色
     */
    getAllroles: function () {
        let args = {
            tpl: "list",
            url: "/system/core/role/list",
            contentId: "list"
        };
        this.search(args);
    },

    /**
     * 点击资源，取得授权用户，更新用户树
     */
    selResc:function(rescid,rescname){
        if(!rescid){
            $("#"+this.options.rightId+" #btnSure").attr('disabled',true);
        }else{
            $("#"+this.options.rightId+" #btnSure").removeAttr('disabled');
        }
        $("#"+this.options.rightId+" #hidRescid").val(rescid);
        $("#"+this.options.rightId+" #spnCurResc").text(rescname);

        if(rescid){
            var url = "system/core/resource/usersbyrescid/"+rescid;
            $.ajax({
                cache : false,
                async : false,// 同步加载
                type : "GET",
                url : url,
                dataType : "json",
                success : function(data) {
                    var zTree = $.fn.zTree.getZTreeObj(this.options.rightTreeId);
                    zTree.checkAllNodes(false); //用户数选择状态初始化
                    if(data.data.length>0){
                       $.each( data.data,function(i,item) {
                            var node = zTree.getNodeByParam("id",item.id);
                            if(node){
                                node.checked=true;
//										zTree.updateNode(node,true);//父节点级联选中
                                zTree.updateNode(node,true);//父节点级联选中
                                zTree.expandNode(node.getParentNode(), true, false, false);  //展开选中节点
                            }

                        }.bind(this));
                    }
                }.bind(this)
            });
        }
    },


//     searchRoleList: function () {
//         var data = $("#" + this.options.searchFormId).serialize();
//         Utils.ajaxByPost({
//             url: this.options.roleListUrl,
//             tplUrl: 'system/core/authorize/rolelistbind',
//             contentId: this.options.roleListId,
//             data: data,
//             calback: function (data) {
//                 //没有数据时处理最大页数
//                 var maxpage = parseInt(data.data.totalPages);
//                 var curpage = parseInt(data.data.number) + 1;
//                 if (maxpage == 0) {
//                     maxpage = 1;
//                 }
//                 //回写最大页数
//                 $(".pagination").jqPagination("option", {
//                     max_page: maxpage,
//                     current_page: curpage,
//                     trigger: false
//                 });
//
//                 $('#' + this.options.roleListId + "input").iCheck({
//                     handle: 'checkbox',
//                     checkboxClass: 'icheckbox_square-blue',
//                     radioClass: 'iradio_square-blue',
//                     increaseArea: '20%' // optional
//                 });
//
//
//                 var rolechk = $('#' + this.options.roleListId + " input:checkbox:not([id='selectAll'])");
//                 var main = this;
//                 //显示资源的绑定角色
//                 if (main.options.map.size() > 0) {
//                     //设置选中用户的状态
//                     rolechk.each(function () {
//                         if (main.options.map.containsKey($(this).attr('objid'))) {
//                             $(this).iCheck('check');
//                         }
//                     });
//                 }
//                 //添加checkbox的选中事件，如果选中则加入缓存中
//                 rolechk.on('ifChecked', function (event) {
//                     var chk = event.target;
//                     main.options.map.put($(chk).attr('objid'), $(chk).attr('objname'));
//                 });
//                 //添加checkbox的选中事件，如果选中则加入缓存中
//                 rolechk.on('ifUnchecked', function (event) {
//                     var chk = event.target;
//                     main.options.map.remove($(chk).attr('objid'));
//                 });
//                 //全选操作
//                 var selectAllChk = $('#' + this.options.roleListId + " input:checkbox#selectAll");
//                 selectAllChk.on('ifChecked', function (event) {
//                     rolechk.iCheck('check');
// //						  rolechk.each(function(){
// //							  main.options.map.put($(this).attr('objid'),$(this).attr('objname'));
// //						  });
//
//                 });
//                 selectAllChk.on('ifUnchecked', function (event) {
//                     rolechk.iCheck('uncheck');
// //						  rolechk.each(function(){
// //							  main.options.map.remove($(this).attr('objid'));
// //						  });
//                 });
//             }.bind(this)
//         });
//     },


    searchNodes: function () {

    },


    /**
     * ------------------------------------zTree上部的查询框和查询----------------------------------
     * TODO 获取选择的节点
     * TODO 查询全部的角色
     * TODO 查询这个节点的角色
     */

    /**
     * zTree上部的查询框
     * @param element
     */
    searchZtree: function (element) {
        var treeId = this.options.leftTreeId;
        var searchConditionId = "txtSearch";
        var searchCondition = $('#' + searchConditionId).val();
        var searchObj = {};
        searchObj.treeId = treeId;
        searchObj.type = 'resource,bizsys';
        searchObj.condition = searchCondition;
        //获得选择的那个节点
        var nodeObj = this.treeSearchAndExpand(searchObj);
        if (nodeObj&&nodeObj.id) {
            this.selResc(nodeObj.id, nodeObj.name);
            //this.selUser(nodeObj.id, nodeObj.name,nodeObj.approvalType);

        }else{
           layer.msg(`没有查询到指定的资源！`);
        }
    },

    /**
     * 树状列表查询后展示方法
     *
     * @param
     * searchObj,查询条件对象，内部含有treeId和condition属性。treeId树ID，condition是查询条件，树状列表上部查询框提供的节点名称查询条件
     * @param treeId树ID
     * @returns
     */
    treeSearchAndExpand: function (searchObj) {
        var treeObj = $.fn.zTree.getZTreeObj(searchObj.treeId);
        // <1>. 先把全部节点更新为普通样式
        var treeNodes = treeObj.transformToArray(treeObj.getNodes());
        for (var i = 0; i < treeNodes.length; i++) {
            treeNodes[i].highlight = false;
            treeObj.updateNode(treeNodes[i]);
        }
        // <2>.得到模糊匹配搜索条件的节点数组集合
        var highlightNodes = new Array();

        if (searchObj.condition != "") {
            highlightNodes = treeObj.getNodesByParamFuzzy("name", searchObj.condition, null);
        } else {
            return;
        }
        // <3>.高亮显示并展示【指定节点s】
        var nodeObj = {};
        if (highlightNodes != null && highlightNodes.length > 0) {
            treeObj.cancelSelectedNode();
            var first = true;
            for (var i = 0; i < highlightNodes.length; i++) {
                if (searchObj.type && searchObj.type.indexOf(highlightNodes[i].type) < 0) {
                    continue;
                }
                // treeObj.selectNode(highlightNodes[i], true);
                highlightNodes[i].highlight = true;
                if (first) {
                    nodeObj = highlightNodes[i];
                    treeObj.selectNode(highlightNodes[i], true);
                    first = false;
                }
                treeObj.updateNode(highlightNodes[i]);// 级联选中
                treeObj.expandNode(highlightNodes[i].getParentNode(), true);
            }
        }
        return nodeObj;
    },
    /**
     * 保存资源用户绑定信息
     */
    bindUsers:function(){
        //右侧选中用户
        var treeObj = $.fn.zTree.getZTreeObj(this.options.rightTreeId);
        var chknodes = treeObj.getCheckedNodes(true);
        var bindnodes="";
        $.each(chknodes,function( i,item) {
            if(item.type=='user'){
                bindnodes+= item.id+",";
            }

        });
        //左侧选中资源
        var rescid= $("#"+this.options.rightId+" #hidRescid").val();
        if(rescid.length==0 ){
            //TODO:替换为标准提示框
           layer.msg(`请选择绑定的资源`);
            //alert('');
            return;
        }
        var param = {"rescid":rescid,userids:bindnodes};
        var url="system/core/authorize/bindrescusers";
        $.ajax({
            cache : false,
            data:param,
            type : "POST",
            url : url,
            dataType : "json",
            success : function(data) {
            }.bind(this)
        });
//			var url="core/resource/bindusers/"+rescid+"/"+bindnodes;
//			$.ajax({
//				cache : false,
//				type : "GET",
//				url : url,
//				dataType : "json",
//				success : function(data) {
//
//				}.bind(this)
//			});
    },
    bindReset:function(){
//			//默认选中第一个资源
        var rescid= $("#"+this.options.rightId+" #hidRescid").val();
        var rescname= $("#"+this.options.rightId+" #spnCurResc").text();
        //显示用户的绑定角色
        this.selResc(rescid, rescname);
    }
});
