// ezdev-pcview/layout/navbar/message/warn/WarnNav.js
/**
 * Created by zhaofs on 2016/9/12.
 * --------------------
 * - 导航菜单 -
 * --------------------
 * 基本导航菜单
 */
import BaseMessage from "../BaseMessage";
import renderer from "../../../../render/Renderer";

export default class extends BaseMessage {
    constructor(options) {
        super(Object.assign({
            name: "nav/warn",
            order: 30
        }, options));
    }

    resolveTpl(tpl) {
        return require("./tpl/" + tpl + ".html");
    }

    /**
     * 生成html
     */
    generateHtml() {
        //当确定登录用户账户后，初始化该用户的“通知”区域，包括“通知”数、未读通知等
        window.ezdev.addEvent("initializedUser", function(user) {
            //初始化“通知”区域
            this.generateWarn(user.account);
        }.bind(this));
        this.addEvent("refreshWarn", function(userAccount) {
            //初始化“通知”区域
            this.generateWarn(userAccount);
        }.bind(this));
        return this.resolveTpl("index")({});
    }

    //初始化“告警”区域
    //本方法用于第一次渲染页面时，手动抓取用户的“告警”
    generateWarn(userid){
        //获取当前用户的未读消息条数
        var argsGetCount = {
            url:"/message/getUnconsumedMessageCount"
            , type:"post"
            , data:"userid="+userid+"&topic=warn"
        };
        this.ajaxResource(argsGetCount).then(
            function (dataCount) {
                //获取当前用户最后5条未读消息
                var argsGetLastMessage = {
                    url:"/message/getLastUnconsumedMessage"
                    , type:"post"
                    , data:"userid="+userid+"&count=10&topic=warn"
                };
                this.ajaxResource(argsGetLastMessage).then(
                    function (dataMessage) {

                        var data = {
                            msgCount:dataCount.data,
                            msgs:dataMessage.data
                        };
                        //填充到模板html内
                        $(".dropdown.warn-nav").replaceWith(this.resolveTpl("index")(data));
                        renderer.analyzeHtml($(".dropdown.warn-nav"));
                    }.bind(this)
                );
            }.bind(this)
        );
    }

    /**
     * 渲染完成后对html逻辑处理,根据需求决定是否需要重载
     * @param container 导航栏容器
     */
    disposeHtml(container) {

        //本方法用于新告警到达时，自动根据新通知内容更新用户的“告警”区域
        //收到“告警”消息后，对“告警”区域进行重绘
        window.ezdev.addEvent('warnMsgArrived',function (msg) {

            var warn = undefined;
            var warnCount = 0;
            var warnUnconsumed = {};

            //对消息的过滤
            for(var i=0; i<msg.pms.length; i++){
                var aMsg = msg.pms[i];
                if(aMsg.title.indexOf("warn") == 0){
                    warn = aMsg.content;
                }
                if(aMsg.title.indexOf("warnCount") == 0){
                    warnCount = aMsg.content;
                }
                if(aMsg.title.indexOf("warnUnconsumed") == 0){
                    warnUnconsumed = aMsg.content;
                }
            }

            var data = {
                msgCount:warnCount,
                msgs:warnUnconsumed
            };

            //填充到模板html内
            $(".dropdown.warn-nav").replaceWith(this.resolveTpl("index")(data));
            renderer.analyzeHtml($(".dropdown.warn-nav"));
        }.bind(this));
    }

    doClick(args){
        this.fireEvent("warnClicked",args.data);
        this.fireEvent("warnClickedId",args.dataId);
    }

    doAllClick(){
        this.fireEvent("warnAllClicked");
    }
};
